<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
<head>
 <META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">
 <META NAME="Generator" CONTENT="Hacker Hands">
 <META NAME="author" CONTENT="Josep F. Abril">
  <!-- $Id: index.html,v 1.1 2003-06-05 17:47:51 jabril Exp $ -->
  <META NAME="keywords"    CONTENT="GFF2APLOT HOWTOs, tutorials, GFF, gff-format, Genomic Pairwise Alignments PostScript Plots, Genomic Analysis, Sinteny, Bioinformatics">
  <META NAME="description" CONTENT="GFF2APLOT HOWTOs: First Contact.">
 <TITLE>GFF2APLOT HOWTOs: First Contact.</TITLE>
<!-- <BASE HREF="http://genome.imim.es/software/gfftools/gff2aplot_docs/first_contact/"> -->
<link rel=stylesheet type="text/css" href="http://genome.imim.es/Genome.css" title="Genome">
<link rel=stylesheet type="text/css" href="http://genome.imim.es/software/gfftools/gff2aplot_docs/HowTo.css" title="Genome">
<STYLE type="text/css">
 @import url(http://genome.imim.es/Genome_forms_mozilla.css);
</STYLE>
<SCRIPT type="text/javascript" src="http://genome.imim.es/Genome.js"></SCRIPT>
</head>

<body onResize="init();" MARGINWIDTH="2" MARGINHEIGHT="0">
<A NAME="TOP"></A>

<!-- ############### TITLE AREA ############### -->
<table border=0 cellpadding=0 cellspacing=0 width="100%">
<tr>
<td align="center">
<FONT size=4 class="tit">Genome BioInformatics Research Lab</FONT>
</td>
</tr>
</table>

<!-- ############### MENU AREA ############### -->
<table class="default" border=0 cellpadding=0 cellspacing=0 width="100%">
<tr>
<td class="links" align="CENTER">
 <a href="http://genome.imim.es/main/help.html" 
  onMouseover="window.status='Asking for Help';">Help</a>
&nbsp;<FONT style="color: #005050;">&#124;</FONT>&nbsp;
 <a href="http://genome.imim.es/main/news.html"
  onMouseover="window.status='News';">News</a>
&nbsp;<FONT style="color: #005050;">&#124;</FONT>&nbsp;
 <a href="http://genome.imim.es/main/people.html"
  onMouseover="window.status='Who is who in Genome Informatics Research Laboratory';">People</a>
&nbsp;<FONT style="color: #005050;">&#124;</FONT>&nbsp;
 <a href="http://genome.imim.es/main/research.html"
  onMouseover="window.status='Research at our Group';">Research</a>
&nbsp;<FONT style="background: #005050; color: #FFFFFF;">&nbsp;
 <a href="http://genome.imim.es/main/software.html"
  style="background: #005050; color: #DDDDDD;"
  onMouseover="window.status='Software developed in our Group';">Software</a>
&nbsp;</FONT>&nbsp; <!-- Current menu -->
 <a href="http://genome.imim.es/main/publications.html"
  onMouseover="window.status='Publications by our Group';">Publications</a>
&nbsp;<FONT style="color: #005050;">&#124;</FONT>&nbsp;
 <a href="http://genome.imim.es/main/links.html"
  onMouseover="window.status='Interesting Links';">Links</a>
</td>
</tr>
<tr>
<td class="links" align="CENTER">
 <a href="http://genome.imim.es/main/databases.html"
  onMouseover="window.status='Datasets created/curated by our Group';" class="hh">Resources &amp; Datasets</a>
&nbsp;<FONT style="color: #005050;">&#124;</FONT>&nbsp;
 <a href="http://genome.imim.es/genepredictions/index.html"
  onMouseover="window.status='Gene Predictions produced by our Software';" class="hh">Gene Predictions</a>
&nbsp;<FONT style="color: #005050;">&#124;</FONT>&nbsp;
 <a href="http://genome.imim.es/main/seminars.html"
  onMouseover="window.status='Seminars, Courses and Group Sessions';" class="hh">Seminars &amp; Courses</a>
</td>
</tr>
</table><BR class="hhh">

<!-- ############### PATH AREA ############### -->
<table border=1 cellpadding=0 cellspacing=0 width="100%">
<tr>
<td class="path" border=0 cellpadding=0 cellspacing=0 ALIGN="LEFT">
 <table border=0 cellpadding=0 cellspacing=1>
 <tr>
 <td class="ppath" border=0 cellpadding=0 cellspacing=0>
&nbsp;
<a href="http://www.imim.es/"
 onMouseover="window.status='Institut Municipal d\'Investigaci&oacute; M&egrave;dica';">IMIM</a>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<IMG SRC="http://genome.imim.es/g_icons/dmd.gif" ALT="*" HEIGHT=8 WIDTH=8 BORDER=0>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<a href="http://www.upf.es/"
 onMouseover="window.status='Universitat Pompeu Fabra';">UPF</a>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<IMG SRC="http://genome.imim.es/g_icons/dmd.gif" ALT="*" HEIGHT=8 WIDTH=8 BORDER=0>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<a href="http://www.crg.es/"
 onMouseover="window.status='Centre de Regulaci&oacute; Gen&ograve;mica';">CRG</a>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<IMG SRC="http://genome.imim.es/g_icons/dmd.gif" ALT="*" HEIGHT=8 WIDTH=8 BORDER=0>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<a href="http://www.imim.es/grib/eng/default.htm"
 onMouseover="window.status='Grup de Recerca en Informàtica Biom&egrave;dica';">GRIB</a>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<A HREF="http://genome.imim.es/index.html" onMouseover="window.status='Genome Informatics Research Laboratory HOME PAGE: You are Welcome!!!'">
<IMG SRC="http://genome.imim.es/g_icons/lhr.gif" ALT="HOME" HEIGHT=10 WIDTH=36 BORDER=0>
</A>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<a href="http://genome.imim.es/main/software.html"
  onMouseover="window.status='Software';">SOFTWARE</a>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<IMG SRC="http://genome.imim.es/g_icons/dmd.gif" ALT="*" HEIGHT=8 WIDTH=8 BORDER=0>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<a href="http://genome.imim.es/software/gfftools/GFF2APLOT.html#TOP"
 onMouseover="window.status='This is the main page for GFF2APLOT.';">gff2aplot</a>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<IMG SRC="http://genome.imim.es/g_icons/dmd.gif" ALT="*" HEIGHT=8 WIDTH=8 BORDER=0>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<a href="http://genome.imim.es/software/gfftools/GFF2APLOT.html#HOWTOS"
 onMouseover="window.status='GFF2APLOT Tutorials...';">HowTo</a>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<IMG SRC="http://genome.imim.es/g_icons/dmd.gif" ALT="*" HEIGHT=8 WIDTH=8 BORDER=0>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<a href="http://genome.imim.es/software/gfftools/gff2aplot_docs/first_contact/index.html#TOP"
 onMouseover="window.status='Getting introduced to GFF2APLOT';">First Contact</a>
 </td>
 </tr>
 </table>
</td>
</tr>
</table>

<!-- ############### USERS AREA ############### -->
<table border=0 cellpadding=0 cellspacing=0 width="100%">
<tr border=0 cellpadding=0 cellspacing=0>
<td class="section" border=0 cellpadding=0 cellspacing=0 width="10px">&nbsp;</td>
<td class="section" ALIGN=CENTER>
 <CENTER><FONT size=6 class="tgen">Introduction to <code>gff2aplot</code></FONT></CENTER>
</td>
<td class="section" border=0 cellpadding=0 cellspacing=0 width="10px">&nbsp;</td>
</tr>
<tr border=0 cellpadding=0 cellspacing=0>
<td class="userspace" border=0 cellpadding=0 cellspacing=0 width="10px">&nbsp;</td>
<td class="userspace" border=0 cellpadding=0 cellspacing=0>

&nbsp;<BR>
<A NAME="TOC"></A>
<P>
<FONT size=5 class="K"> Summary </FONT> <BR class="kk"><BR>
The main goal of this tutorial is to introduce you on the basics of using <code>gff2aplot</code>. We will see how to run the program on a unix shell and few command-line switches. After that, we will introduce the customization files and the basics of customizing the program's output. Then we will look at examples of the GFF input records. Finally, we will build up a figure following three easy steps.  <BR class="hh"><BR>
<B>NOTE.-</B> For the sake of clarity, we are going to use long names for the comand-line switches. See the command-line help if you prefer short names for those cases in which a short name is available. <BR>
Bitmaps for the examples were generated as PNGs (Portable Network Graphics). If your browser is not ready for such format yet, you can visualize the PDF or PS versions by clicking on the links below each snapshot. Links to customization files, log files, GFF input files and output PostScript figures, are also available on each command-line shown. <BR>
</P>
<P>
<FONT size=5 class="K"> Contents </FONT>
<UL>
  <LI><A HREF="index.html#BSCL" onMouseover="window.status='Basic command-line switches';">Basic command-line switches.</A> <BR></LI>
  <LI><A HREF="index.html#CUSTOM" onMouseover="window.status='Customization files';">Customization files.</A> <BR></LI>
  <LI><A HREF="index.html#GFF" onMouseover="window.status='GFF FORMAT';">Understanding GFF records.</A> <BR></LI>
  <LI><A HREF="index.html#EXCL" onMouseover="window.status='Three steps example';">Running <code>gff2aplot</code>.</A> <BR></LI>
</UL> <BR class="hh">
</P>

<TABLE border=0 cellpadding=0 cellspacing=0 width="100%">
<TR>
<TD class="section" width=80% ALIGN=LEFT>
<a name="BSCL"></a><FONT size=5 class="K"> Command-Line Basics </FONT>
</TD>
<TD class="section" width=20% ALIGN=RIGHT>
<a href="index.html#TOP"
 onMouseover="window.status='TOP of the PAGE';">
<IMG class="pnt" SRC="http://genome.imim.es/g_icons/top.gif" HEIGHT=15 WIDTH=15 BORDER=0></a>
</TD></TR>
</TABLE>
<P> <BR class="hh">
From now on, we will represent bash shell command-line prompt as <CODE class="rtbl">&nbsp;<B>$></B>&nbsp;</CODE>. The most basic command-line switches are those to retrieve info regarding program execution parameters and current version. <BR>
</P>
<P><TABLE class="userspace" border=2 cellpadding=0 cellspacing=0>
<TR><TD class="rtbl">&nbsp;<BR class="hh">&nbsp;&nbsp;<B>$></B><BR class="hh"></TD><TD class="codebg" ALIGN=LEFT>&nbsp;<BR class="hh"><B><PRE class="codefg" style="font-size: 10px;">
 gff2aplot.pl --help <BR class="hh">
</PRE></B></TD><TD class="codebg">&nbsp;&nbsp;</TD></TR>
<TR><TD class="rtbl">&nbsp;&nbsp;<B>$></B><BR></TD><TD class="codebg" ALIGN=LEFT><B><PRE class="codefg" style="font-size: 10px;">
 gff2aplot.pl --version <BR class="hh">
</PRE></B></TD><TD class="codebg">&nbsp;&nbsp;</TD></TR>
</TABLE></P>
<P>
You can take a look to the output produced by those two command-line switches from these links: <a href="gff2aplot.help"><CODE><B>--help</B></CODE> output</a> and <a href="gff2aplot.help"><CODE><B>--verbose</B></CODE> output</a>. <BR class="hh"><BR>
We must think of <code>gff2aplot</code> as a Unix filter, as it can process input GFF records from standard input and/or from GFF files. In the first case, data can be read through a command-line Unix pipeline (<CODE style="background: #DDFFDD;">&nbsp;command | gff2aplot.pl&nbsp;</CODE>) or from a file redirection (<CODE style="background: #DDFFDD;">&nbsp;gff2aplot.pl < input_file.gff&nbsp;</CODE>). Some examples of this command-line flexibility are shown below. <BR>
</P>
<P><TABLE class="userspace" border=2 cellpadding=0 cellspacing=0>
<TR><TD class="rtbl">&nbsp;<BR class="hh">&nbsp;&nbsp;<B>$></B><BR></TD><TD class="codebg" ALIGN=LEFT>&nbsp;<BR class="hh"><B><PRE class="codefg" style="font-size: 10px;">
 gff2aplot.pl --quiet [ ...more options... -- ] input_file.gff > output_file.ps 2> report_file.err <BR class="hh">
</PRE></B></TD><TD class="codebg">&nbsp;&nbsp;</TD></TR>
</TABLE></P>
<P>
Previous program execution took GFF records from a file (<CODE>input_file.gff</CODE>). <CODE><B>--quiet</B></CODE> switch disables any message to standard error except those fatal errors that abort program normal progress for instance. <BR>
</P>
<P><TABLE class="userspace" border=2 cellpadding=0 cellspacing=0>
<TR><TD class="rtbl">&nbsp;<BR class="hh">&nbsp;&nbsp;<B>$></B><BR></TD><TD class="codebg" ALIGN=LEFT>&nbsp;<BR class="hh"><B><PRE class="codefg" style="font-size: 10px;">
 gff2aplot.pl --verbose [ ...more options... -- ] <B><</B> input_file.gff > output_file.ps 2> report_file.log+err <BR class="hh">
</PRE></B></TD><TD class="codebg">&nbsp;&nbsp;</TD></TR>
</TABLE></P>
<P>
On the last command-line, we read GFF records from standard input provided via filehandle redirection. <CODE><B>--verbose</B></CODE> shows the program execution progress report via standard error. <BR>
</P>
<P><TABLE class="userspace" border=2 cellpadding=0 cellspacing=0>
<TR><TD class="rtbl">&nbsp;<BR class="hh">&nbsp;&nbsp;<B>$></B><BR></TD><TD class="codebg" ALIGN=LEFT>&nbsp;<BR class="hh"><B><PRE class="codefg" style="font-size: 10px;">
 cat input_file.gff <B>|</B> gff2aplot.pl [ ...more options... -- ] - > output_file.ps 2> report_file.log+err <BR class="hh">
</PRE></B></TD><TD class="codebg">&nbsp;&nbsp;</TD></TR>
</TABLE></P>
<P>
Shown above is input parsing of GFF records from a unix pipeline. As no verbose switch was provided only warnings are shown through standard error. <BR> 
</P>
<P><TABLE class="userspace" border=2 cellpadding=0 cellspacing=0>
<TR><TD class="rtbl">&nbsp;<BR class="hh">&nbsp;&nbsp;<B>$></B><BR></TD><TD class="codebg" ALIGN=LEFT>&nbsp;<BR class="hh"><B><PRE class="codefg" style="font-size: 10px;">
 gff2aplot.pl --logs-filename report_file.log [ ...more options... -- ] < input_file.gff > output_file.ps 2> report_file.err <BR class="hh">
</PRE></B></TD><TD class="codebg">&nbsp;&nbsp;</TD></TR>
</TABLE></P>
<P>
<CODE><B>--logs-filename</B></CODE> enables verbose mode too, but redirecting the progres report to the given report file. Warnings are sent both to the report file and to standard error. <BR class="hh"><BR>
Here you can find a more complex example of command-line: <BR>
</P>
<P><TABLE class="userspace" border=2 cellpadding=0 cellspacing=0>
<TR><TD class="rtbl">&nbsp;<BR class="hh">&nbsp;&nbsp;<B>$></B><BR></TD><TD class="codebg" ALIGN=LEFT>&nbsp;<BR class="hh"><B><PRE class="codefg" style="font-size: 10px;">
gff2aplot.pl                                      \
    --custom-filename file_1.rc                   \
    --custom-filename file_2.rc                   \
    --title 'CMD-LINE USAGE'                      \
    --subtitle                                    \
      'This is an example on command-line usage.' \
    --x-sequence-zoom  100..400                   \
    --y-sequence-zoom  600..900                   \
    --zoom-area                                   \
    --feature-var '*::alignment_color=darkblue'   \
    -- annotation.gff alignment.gff               \
     > output_file.ps <BR class="hh">
</PRE></B></TD><TD class="codebg">&nbsp;&nbsp;</TD></TR>
</TABLE></P>
<P>
If verbose mode is enabled, when the previous command-line options and file names are being validated, <code>gff2aplot</code> will report something resembling the following output: <BR>
</P>
<P><TABLE class="userspace" border=2 cellpadding=0 cellspacing=0>
<TR><TD class="ytbl">&nbsp;&nbsp;</TD><TD class="ytbl" ALIGN=LEFT>&nbsp;<BR class="hh"><B><PRE class="codefg" style="font-size: 10px;">
  &#183;&#183;&#183; 
################################################################################
###                             SETTING DEFAULTS                             ###
################################################################################
################################################################################
###                      CHECKING COMMAND-LINE OPTIONS                       ###
################################################################################
###
### Validating INPUT FILENAMES
###
###---> "annotation.gff" exists, including as Input File.
###---> "alignment.gff" exists, including as Input File.
###
### Validating CUSTOM FILENAMES
###
###---> ".gff2aplotrc" NOT FOUND, skipping!!!
###---> "file_1.rc" exists, including as Custom File.
###---> "file_2.rc" exists, including as Custom File.
###
### Checking COMMAND-LINE Options
###
..... [000005]
###
### Checking Custom Variables SET by COMMAND-LINE
###
F [000001]
################################################################################
###                       COMMAND-LINE OPTIONS CHECKED                       ###
###           0 wallclock secs ( 0.06 usr +  0.00 sys =  0.06 CPU)           ###
################################################################################
  &#183;&#183;&#183;  <BR class="hh">
</PRE></B></TD><TD class="ytbl">&nbsp;&nbsp;</TD></TR>
</TABLE></P>
<P>
Imagine the following scenario, you are waiting for <code>gff2aplot</code> to finish, but you are not providing input records via the standard input chanel. You will wait forever that <code>gff2aplot</code> will finish. Take that into account when passing data through standard input. If you stop porgram execution (i.e. pressing <CODE>CTRL-C</CODE>) you will be warned with a message like the following: <BR>
</P>
<P><TABLE class="userspace" border=2 cellpadding=0 cellspacing=0>
<TR><TD class="ytbl">&nbsp;&nbsp;</TD><TD class="ytbl" ALIGN=LEFT>&nbsp;<BR class="hh"><B><PRE class="codefg" style="font-size: 10px;">
<<<--------->>>
<<< WARNING >>> gff2aplot has been stopped by user !!!
<<<--------->>>
<<< WARNING >>> ---------- Exiting NOW !!! ----------
<<<--------->>> <BR class="hh">
</PRE></B></TD><TD class="ytbl">&nbsp;&nbsp;</TD></TR>
</TABLE></P>

<TABLE border=0 cellpadding=0 cellspacing=0 width="100%">
<TR>
<TD class="section" width=80% ALIGN=LEFT>
<a name="CUSTOM"></a><FONT size=5 class="K"> Customization Files </FONT>
</TD>
<TD class="section" width=20% ALIGN=RIGHT>
<a href="index.html#TOP"
 onMouseover="window.status='TOP of the PAGE';">
<IMG class="pnt" SRC="http://genome.imim.es/g_icons/top.gif" HEIGHT=15 WIDTH=15 BORDER=0></a>
</TD></TR>
</TABLE>
<P> <BR class="hh">
The following table summarizes few aspects of a <code>gff2aplot</code> customization file: <BR>
</P>
<P><TABLE class="userspace" border=2 cellpadding=0 cellspacing=0>
<TR><TD class="rtbl">&nbsp;&nbsp;</TD><TD class="rtbl" ALIGN=LEFT>&nbsp;<BR class="hh"><B><PRE class="codefg" style="font-size: 10px;">
# 
# Optional Header
#                           (<b>i.e. program version, creation date, user id,</b>
#                                         <b>version control tags, and so on</b>)
#  
#                                   (<b>Comments and empty lines are skipped</b>) 

# L #                                 (<b>This is Block Separator for Layout</b>) 

variable_name<sub><small>1</small></sub>=value<font style="background: #FFFFFF;font-size: 5pt;">&nbsp;&nbsp;</font># blah,blah,blah&#133; 
#              (<b>Comments are also allowed after the value definition</b>
#               <b>Note that there must be a single blank space at least,</b>
#               <b>shown here as '<font style="background: #FFFFFF;font-size: 5pt;">&nbsp;&nbsp;</font>' before #. Otherwise, the comment</b>
#               <b>will be considered part of the variable value</b>) 

  &#183;&#183;&#183; 
# blah, blah, blah&#133;    (<b>Extra comment-lines can be added where you need</b>) 

  &#183;&#183;&#183; 
variable_name<sub><small>n</small></sub>=value 

# F # blah, blah, blah&#133; 
#     (<b>You can place extra comments also on section headers, after second #</b>) 

GFF-feature_key<sub><small>1</small></sub>::feature_variable_name=value 
  &#183;&#183;&#183; 
GFF-feature_key<sub><small>n</small></sub>::feature_variable_name=value 

# G # 
GFF-group_key<sub><small>1</small></sub>::group_variable_name=value 
  &#183;&#183;&#183; 
GFF-group_key<sub><small>n</small></sub>::group_variable_name=value 

# S # 
GFF-source_key<sub><small>1</small></sub>::source_variable_name=value 
  &#183;&#183;&#183; 
GFF-source_key<sub><small>n</small></sub>::source_variable_name=value

# X #
# Examples of features available for this section
# are available at the <A HREF="http://genome.imim.es/software/gfftools/gff2aplot_docs/plot_layers/index.html#PRE">"Layers on <code>gff2aplot</code> Figures" tutorial</A> <BR>
</PRE></B></TD><TD class="rtbl">&nbsp;&nbsp;</TD></TR>
</TABLE></P>
<P>
One important feature of such files is that they are not only optional but also any variable definition within them, even any of the file sections too. So that, you can start up with a simple customization file containing just few settings and grow it up as you need. As you can now set any of those customization parameters from command-line (via the <CODE><B>--layout-var</B></CODE>, <CODE><B>--source-var</B></CODE>, <CODE><B>--group-var</B></CODE>, <CODE><B>--feature-var</B></CODE>, ..., command-line switches), you are able to test them before including a new variable into the customization file. You may find examples of this on several <code>gff2aplot</code> tutorials. <BR>
You can also provide more than one customization file. You must take into account that when a variable appears more than once, latest uploaded value will be the one finally set (later definitions override previous ones). Furthermore, any command-line switch will override the corresponding variable value. <BR>
The customization files are modular. This means that there is a set of variables that can be used within a section, but those sections can be placed in any order, or they may appear several times in a customization file. When parsing those files, the program will classify each variable to create a table with all the settings. The sections must begin with a section header, which is a special comment line starting with five fixed chars. That string will follow the pattern '<CODE><B># * #</B></CODE>': a hash mark followed by a white space, followed by the section code char, then another white space and a closing hash mark. If verbose mode is on, program will output the following report for each of the custom files parsed: <BR>
</P>
<P><TABLE class="userspace" border=2 cellpadding=0 cellspacing=0>
<TR><TD class="ytbl">&nbsp;&nbsp;</TD><TD class="ytbl" ALIGN=LEFT>&nbsp;<BR class="hh"><B><PRE class="codefg" style="font-size: 10px;">
  &#183;&#183;&#183; 
################################################################################
###                           READING CUSTOM FILES                           ###
################################################################################
###
### Reading Customization Parameters from "<A HREF="file_1.rc">file_1.rc</A>"
###
.....*LLLLLLLLLLL..*FFF [000023]
###
### Reading Customization Parameters from "<A HREF="file_2.rc">file_2.rc</A>"
###
......*.LLLLL. [000014]
  &#183;&#183;&#183; 
    (... as many custom files you have passed 
         via commad-line option <CODE><B>--custom-filename <I>file.rc</I></B></CODE> ...)
  &#183;&#183;&#183; 
################################################################################
###                           CUSTOM FILES LOADED                            ###
###           0 wallclock secs ( 0.02 usr +  0.00 sys =  0.02 CPU)           ###
################################################################################
  &#183;&#183;&#183; <BR class="hh">
</PRE></B></TD><TD class="ytbl">&nbsp;&nbsp;</TD></TR>
</TABLE></P>
<P>
From that example we found that <A HREF="file_1.rc"><B><I>file_1.rc</I></B> has 23 lines</A>: 7 empty records and comments, 2 sections, 11 layout variable definitions and settings for 3 feature vars. The following table summarizes the codes used to describe what is being processed from customization files. <BR>
</P>
<P><TABLE class="userspace" border=2 cellpadding=0 cellspacing=0>
<TR><TH class="rtbl" ALIGN=CENTER>&nbsp;&nbsp;Record Type&nbsp;&nbsp;</TH><TH class="rtbl" ALIGN=CENTER>&nbsp;&nbsp;Symbol Used&nbsp;&nbsp;</TH></TR>
<TR><TD class="rtbl" ALIGN=CENTER>Empty Records<BR class="hh">or Comments</TD><TD class="ytbl" ALIGN=CENTER><B>.</B></TD></TR>
<TR><TD class="rtbl" ALIGN=CENTER>Section Header</TD><TD class="ytbl" ALIGN=CENTER><B>*</B></TD></TR>
<TR><TD class="rtbl" ALIGN=CENTER>   Layout Var </TD><TD class="ytbl" ALIGN=CENTER>L</TD></TR>
<TR><TD class="rtbl" ALIGN=CENTER> Sequence Var </TD><TD class="ytbl" ALIGN=CENTER>Q</TD></TR>
<TR><TD class="rtbl" ALIGN=CENTER>   Source Var </TD><TD class="ytbl" ALIGN=CENTER>S</TD></TR>
<TR><TD class="rtbl" ALIGN=CENTER>   Strand Var </TD><TD class="ytbl" ALIGN=CENTER>T</TD></TR>
<TR><TD class="rtbl" ALIGN=CENTER>    Group Var </TD><TD class="ytbl" ALIGN=CENTER>G</TD></TR>
<TR><TD class="rtbl" ALIGN=CENTER>  Feature Var </TD><TD class="ytbl" ALIGN=CENTER>F</TD></TR>
<TR><TD class="rtbl" ALIGN=CENTER> Special Feat </TD><TD class="ytbl" ALIGN=CENTER>X</TD></TR>
<TR><TD class="rtbl" ALIGN=CENTER>    Unknown   </TD><TD class="ytbl" ALIGN=CENTER><B>?</B></TD></TR>
</TABLE></P>

<TABLE border=0 cellpadding=0 cellspacing=0 width="100%">
<TR>
<TD class="section" width=80% ALIGN=LEFT>
<a name="GFF"></a><FONT size=5 class="K"> GFF Records </FONT>
</TD>
<TD class="section" width=20% ALIGN=RIGHT>
<a href="index.html#TOP"
 onMouseover="window.status='TOP of the PAGE';">
<IMG class="pnt" SRC="http://genome.imim.es/g_icons/top.gif" HEIGHT=15 WIDTH=15 BORDER=0></a>
</TD></TR>
</TABLE>
<P> <BR class="hh">
<code>gff2aplot</code> can parse three basic formats: standard GFF (versions 1 and 2) and APLOT format. The second was derived from GFF to produce more compact input files when processing alignment data. There are two basic data sets, those records that contain axes annotations (f.i. gene structures, exons, and so on) and those coding for the pair-wise alignment coordinates. Cases A and B from the table below, are suitable for annotations. Program assumes that cases C, D and E correspond to alignment records.
</P>
<P><CENTER><TABLE class="userspace" border=0 cellpadding=0 cellspacing=0>
<TR><TD class="rtbl">&nbsp;&nbsp;</TD><TD class="rtbl" ALIGN=LEFT><PRE class="codefg" style="font-size: 10px;"><B>
seqname  source  feature  start  end  score  strand  frame  [ attributes ] <BR class="hh">
</B></PRE></TD><TD class="rtbl">&nbsp;&nbsp;</TD></TR>
<TR><TD class="codebg">&nbsp;&nbsp;</TD><TD class="codebg" ALIGN=LEFT>&nbsp; <BR class="hh"><PRE class="codefg" style="font-size: 10px;">
#########
<B>### A ### GFF version 1 record:</B> Axes annotation example
#########
NM_009315 chr5 exon 1756 1901 . + . mus_taf6
NM_005641 chr7 exon 7358 7435 . + . hsap_taf6 <BR class="hh">
#########
<B>### B ### GFF version 2 records:</B> Axes annotation example
#########                        (same features as in the above GFF version 1 axes annotation example)
NM_009315	chr5	exon	1756	1901	.	+	.	Gene "mmus_taf6"
NM_005641	chr7	exon	7358	7435	.	+	.	Gene "hsap_taf6" <BR class="hh">
#########
<B>### C ### GFF version 2 records:</B> Similarity region example
#########
NM_009315	BLASTN	hsp	1719	2099	1122	+	.	Target "NM_005641";	Start 7067;	End 7448;	Strand +;	Frame .;    
NM_009315	BLASTN	hsp	5219	5832	1405	+	.	Target "NM_005641";	Start 11597;	End 12218;	Strand +;	Frame .; <BR class="hh">
#########
<B>### D ### GFF version 2 pseudo-records (compact):</B> Similarity region example
#########                                         (same features as in the previous GFF version 2 similarity region example)
NM_009315	BLASTN	hsp	1719	2099	1122	+	.	Target "NM_005641" 7067 7448 + .
NM_009315	BLASTN	hsp	5219	5832	1405	+	.	Target "NM_005641" 11597 12218 + . <BR>
</PRE></TD><TD class="codebg">&nbsp;&nbsp;</TD></TR>
<TR><TD class="userspace">&nbsp;&nbsp;</TD><TD class="userspace" ALIGN=LEFT><P>

</P></TD><TD class="userspace">&nbsp;&nbsp;</TD></TR>
<TR><TD class="rtbl">&nbsp;&nbsp;</TD><TD class="rtbl" ALIGN=LEFT><PRE class="codefg" style="font-size: 10px;"><B>
seqnameX:Y  source  feature  startX:Y  endX:Y  score  strandX[:Y]  frameX[:Y]  [ group[:id] [ ; ...more attributes... ] ] <BR class="hh">
</B></PRE></TD><TD class="rtbl">&nbsp;&nbsp;</TD></TR>
<TR><TD class="codebg">&nbsp;&nbsp;</TD><TD class="codebg" ALIGN=LEFT>&nbsp; <BR class="hh"><PRE class="codefg" style="font-size: 10px;">
#########
<B>### E ### APLOT record:</B> Similarity region example
#########               (same features as in the above GFF version 2 similarity region example)
NM_009315:NM_005641 BLASTN hsp 1719:7067 2099:7448 1122 +:+ .:. 174.4:2
NM_009315:NM_005641 BLASTN hsp 5219:11597 5832:12218 1405 +:+ .:. 216.9:1 <BR>
</PRE></TD><TD class="codebg">&nbsp;&nbsp;</TD></TR>
</TABLE></P>
<P>
If verbose mode was activated, you will obtain a report like the following when parsing GFF files: <BR>
</P>
<P><TABLE class="userspace" border=2 cellpadding=0 cellspacing=0>
<TR><TD class="ytbl">&nbsp;&nbsp;</TD><TD class="ytbl" ALIGN=LEFT>&nbsp;<BR class="hh"><B><PRE class="codefg" style="font-size: 10px;">
  &#183;&#183;&#183; 
################################################################################
###                        PARSING INPUT GFF RECORDS                         ###
################################################################################
###
### Reading GFF records from "<A HREF="blastn_aplot.gff">blastn_aplot.gff</A>"
###
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO [000042]
###
### Reading GFF records from "<A HREF="mouse.gff">mouse.gff</A>"
###
XXXXXXXXXXXXXXXX [000016]
###
### Reading GFF records from "<A HREF="human.gff">human.gff</A>"
###
XXXXXXXXXXXXXXXXX [000017]
################################################################################
###                               DATA LOADED                                ###
###           0 wallclock secs ( 0.07 usr +  0.02 sys =  0.09 CPU)           ###
################################################################################
  &#183;&#183;&#183; <BR class="hh">
</PRE></B></TD><TD class="ytbl">&nbsp;&nbsp;</TD></TR>
</TABLE></P>
<P>
<code>gff2aplot</code> assigns each GFF record it reads to one of the five classes shown before. The codes that could appear on the above report section are the following: <BR>
</P>
<P><TABLE class="userspace" border=2 cellpadding=0 cellspacing=0>
<TR><TH class="rtbl" ALIGN=CENTER>&nbsp;&nbsp;Record Type&nbsp;&nbsp;</TH><TH class="rtbl" ALIGN=CENTER>&nbsp;&nbsp;Symbol Used&nbsp;&nbsp;</TH></TR>
<TR><TD class="rtbl" ALIGN=CENTER>Empty Records<BR class="hh">or Comments</TD><TD class="ytbl" ALIGN=CENTER><B>.</B></TD></TR>
<TR><TD class="rtbl" ALIGN=CENTER>   GFF (grouped)   </TD><TD class="ytbl" ALIGN=CENTER>X</TD></TR>
<TR><TD class="rtbl" ALIGN=CENTER>  GFF (ungrouped)  </TD><TD class="ytbl" ALIGN=CENTER>x</TD></TR>
<TR><TD class="rtbl" ALIGN=CENTER>  GFF (alignment)  </TD><TD class="ytbl" ALIGN=CENTER>A</TD></TR>
<TR><TD class="rtbl" ALIGN=CENTER>   GFF (vector)    </TD><TD class="ytbl" ALIGN=CENTER>V</TD></TR>
<TR><TD class="rtbl" ALIGN=CENTER>  APLOT (grouped)  </TD><TD class="ytbl" ALIGN=CENTER>O</TD></TR>
<TR><TD class="rtbl" ALIGN=CENTER> APLOT (ungrouped) </TD><TD class="ytbl" ALIGN=CENTER>o</TD></TR>
<TR><TD class="rtbl" ALIGN=CENTER>      Unknown      </TD><TD class="ytbl" ALIGN=CENTER><B>?</B></TD></TR>
</TABLE></P>
<P>
We provide several filters to convert output from different sequence alignment programs into such GFF/APLOT formats. See the distribution README file and specific examples from other <code>gff2aplot</code> tutorials.
</P>

<TABLE border=0 cellpadding=0 cellspacing=0 width="100%">
<TR>
<TD class="section" width=80% ALIGN=LEFT>
<a name="EXCL"></a><FONT size=5 class="K"> Running <code>gff2aplot</code> </FONT>
</TD>
<TD class="section" width=20% ALIGN=RIGHT>
<a href="index.html#TOP"
 onMouseover="window.status='TOP of the PAGE';">
<IMG class="pnt" SRC="http://genome.imim.es/g_icons/top.gif" HEIGHT=15 WIDTH=15 BORDER=0></a>
</TD></TR>
</TABLE>
<P> <BR class="hh">
Let's start this three steps appetizer by letting <code>gff2aplot</code> to assume default values for all the customization parameters. We are going to fed it with a blastn output (filtered with the <code>parseblast</code> script) and the corresponding annotations of the genomic region of the human TAF6 gene and its mouse orthologous gene. As you can see, it is a very simple command-line...
</P>
<P><CENTER><TABLE class="userspace" border=0 cellpadding=0 cellspacing=0><TR><TD class="userspace" ALIGN=CENTER VALIGN=MIDDLE><CENTER><a href="taf6_defaults.png"><IMG SRC="taf6_defaults.s.png" ALT="Default Plot" HEIGHT=150 WIDTH=120 BORDER=1></a> <BR> &nbsp;[<a href="taf6_defaults.png">PNG</a>]&nbsp;[<a href="taf6_defaults.ps">PS</a>]&nbsp;[<a href="taf6_defaults.pdf">PDF</a>]&nbsp; <BR>
</CENTER></TD><TD class="codebg">&nbsp;&nbsp;</TD><TD class="codebg" ALIGN=LEFT>&nbsp; <BR class="hh"><PRE class="codefg">
gff2aplot.pl --verbose             \
             -- <a href="taf6.gff">taf6.gff</a>           \
              > <a href="taf6_defaults.ps">taf6_defaults.ps</a>   \
             2> <a href="taf6_defaults.log">taf6_defaults.log</a> <BR class="hh">
</PRE></TD><TD class="codebg">&nbsp;&nbsp;</TD></TR></TABLE></CENTER></P>
<P>
Now, we can play with several command-line switches to change the resulting PostScript plot according to our needs. For instance, we want to highlight exons by paintind them with a green color. We also want to distinguish the untranslated regions (UTRs), so we choose a blue tone. Adding titles will help to identify what we are showing in this plot. Finally, we would like to see the scores projected along the mouse sequence in the bottom percent panel.
</P>
<P><CENTER><TABLE class="userspace" border=0 cellpadding=0 cellspacing=0><TR><TD class="userspace" ALIGN=CENTER VALIGN=MIDDLE><CENTER><a href="taf6_cmdline.png"><IMG SRC="taf6_cmdline.s.png" ALT="Command-line options" HEIGHT=150 WIDTH=120 BORDER=1></a> <BR> &nbsp;[<a href="taf6_cmdline.png">PNG</a>]&nbsp;[<a href="taf6_cmdline.ps">PS</a>]&nbsp;[<a href="taf6_cmdline.pdf">PDF</a>]&nbsp; <BR>
</CENTER></TD><TD class="codebg">&nbsp;&nbsp;</TD><TD class="codebg" ALIGN=LEFT>&nbsp; <BR class="hh"><PRE class="codefg">
gff2aplot.pl                                                      \
    --verbose                                                     \
    --title 'TAF6 Human/Mouse Orthologous Genes'                  \
    --subtitle                                                    \
      'Figure displays TBLASTX results for this genomic region.'  \
    --show-percent-box                                            \
    --percent-box-label "SIMILARITY"                              \
    --group-var '/.*M.musculus.*/::group_label=taf6 (M.musculus)' \
    --group-var '/.*H.sapiens.*/::group_label=taf6 (H.sapiens)'   \
    --feature-var '/[it].*exon/::feature_color=darkgreen'         \
    --feature-var 'utrexon::feature_color=skyblue'                \
    --feature-var 'initialexon::feature_shape=half_arrow_end'     \
    --feature-var 'terminalexon::feature_shape=half_arrow'        \
    --  <a href="taf6.gff">taf6.gff</a>  > <a href="taf6_cmdline.ps">taf6_cmdline.ps</a>  2> <a href="taf6_cmdline.log">taf6_cmdline.log</a> <BR class="hh">
</PRE></TD><TD class="codebg">&nbsp;&nbsp;</TD></TR></TABLE></CENTER></P>
<P>
Now, we are going to move all those settings into a customization file. This will allow us to reuse the same customization on new data sets to produce a similar figure "layout". Most command-line switches have their counterpart customization file variables, the opposite can be achieved by using the special <code><b>--*-var</b></code> switches. A side effect of moving settings from command-line options into a customization file, is that we are simplifying the command-line...
</P>
<P><CENTER><TABLE class="userspace" border=0 cellpadding=0 cellspacing=0><TR><TD class="userspace" ALIGN=CENTER VALIGN=MIDDLE><CENTER><a href="taf6_final.png"><IMG SRC="taf6_final.s.png" ALT="Using customization file" HEIGHT=150 WIDTH=120 BORDER=1></a> <BR> &nbsp;[<a href="taf6_final.png">PNG</a>]&nbsp;[<a href="taf6_final.ps">PS</a>]&nbsp;[<a href="taf6_final.pdf">PDF</a>]&nbsp; <BR>
</CENTER></TD><TD class="codebg">&nbsp;&nbsp;</TD><TD class="codebg" ALIGN=LEFT>&nbsp; <BR class="hh"><PRE class="codefg">
gff2aplot.pl                                                     \
    --verbose                                                    \
    --title 'TAF6 Human/Mouse Orthologous Genes'                 \
    --subtitle                                                   \
      'Figure displays TBLASTX results for this genomic region.' \
    --custom-filename <a href="taf.rc">taf6.rc</a>                                    \
    --  <a href="taf.gff">taf6.gff</a>  > <a href="taf6_final.ps">taf6_final.ps</a>  2> <a href="taf6_final.log">taf6_final.log</a> <BR class="hh">
</PRE></TD><TD class="codebg">&nbsp;&nbsp;</TD></TR></TABLE></CENTER></P>
<P>
You are ready to start a cycle of testing command-line options, moving those you like into a customization file, and so on. Perhaps you should do that process once, when you start defining a new figure layout. Maybe you will play with all the options or just using defaults, or you will set up several complementary customization files, etcetera. From now on is up to you to explore the features available for <code>gff2aplot</code>. As Perl motto says "<I>There's More Than One Way To Do It</I>"... <BR><BR>
<a href="http://genome.imim.es/software/gfftools/GFF2APLOT.html#HOWTOS" onMouseover="window.status='GFF2APLOT Tutorials...';">Take a look to the available tutorials</a>, perhaps there is one that suits your current plot requests. Thanks for using <code>gff2aplot</code>, we hope you will enjoy it !!! <BR>
</P>

</td>
<td class="userspace" border=0 cellpadding=0 cellspacing=0 width="10px">&nbsp;</td>
</tr>
</table>

<!-- ############### TRAILER AREA ############### -->
<table class="trailer" border=0 cellpadding=0 cellspacing=0 width="100%">
 <tr class="default">
 <td id="traila" align=left>
&nbsp;
<A HREF="http://genome.imim.es/main/disclaimer.html"
 onMouseover="window.status='Data Release Policy and Software License';">Disclaimer</A>
 </td>
 <td id="trailb" align=center>
<SCRIPT type="text/javascript" languaje="javascript">
<!--
 LastUpdateSplit()
// -->
</SCRIPT>
 </td>
 <td id="trailc" align=rigth>
<A HREF="MAILTO:gmaster@imim.es"
 onMouseover="window.status='Contact with our webmaster';">webmaster</A>
&nbsp;
 </td>
 </tr>
</table>

<SCRIPT type="text/javascript" languaje="javascript">
<!--
 reportErrors();
// -->
</SCRIPT>
<!-- ############### EOF ############### -->
</body>
</html>
