<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
<head>
 <META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">
 <META NAME="Generator" CONTENT="Hacker Hands">
 <META NAME="author" CONTENT="Josep F. Abril">
  <!-- $Id: index.html,v 1.2 2003-06-10 10:31:40 jabril Exp $ -->
  <META NAME="keywords"    CONTENT="GFF2APLOT HOWTOs, tutorials, GFF, gff-format, Genomic Pairwise Alignments PostScript Plots, Genomic Analysis, Homology, Sinteny, Bioinformatics, SIM">
  <META NAME="description" CONTENT="GFF2APLOT HOWTOs: Analyzing SIM output for a syntenic sequences pair.">
 <TITLE>GFF2APLOT HOWTOs: SIM output for a syntenic sequences pair.</TITLE>
<!-- <BASE HREF="http://genome.imim.es/software/gfftools/gff2aplot_docs/mhcregion/"> -->
<link rel=stylesheet type="text/css" href="http://genome.imim.es/Genome.css" title="Genome">
<link rel=stylesheet type="text/css" href="http://genome.imim.es/software/gfftools/gff2aplot_docs/HowTo.css" title="Genome">
<STYLE type="text/css">
 @import url(http://genome.imim.es/Genome_forms_mozilla.css);
</STYLE>
<SCRIPT type="text/javascript" src="http://genome.imim.es/Genome.js"></SCRIPT>
</head>

<body onResize="init();" MARGINWIDTH="2" MARGINHEIGHT="0">
<A NAME="TOP"></A>

<!-- ############### TITLE AREA ############### -->
<table border=0 cellpadding=0 cellspacing=0 width="100%">
<tr>
<td align="center">
<FONT size=4 class="tit">Genome BioInformatics Research Lab</FONT>
</td>
</tr>
</table>

<!-- ############### MENU AREA ############### -->
<table class="default" border=0 cellpadding=0 cellspacing=0 width="100%">
<tr>
<td class="links" align="CENTER">
 <a href="http://genome.imim.es/main/help.html" 
  onMouseover="window.status='Asking for Help';">Help</a>
&nbsp;<FONT style="color: #005050;">&#124;</FONT>&nbsp;
 <a href="http://genome.imim.es/main/news.html"
  onMouseover="window.status='News';">News</a>
&nbsp;<FONT style="color: #005050;">&#124;</FONT>&nbsp;
 <a href="http://genome.imim.es/main/people.html"
  onMouseover="window.status='Who is who in Genome Informatics Research Laboratory';">People</a>
&nbsp;<FONT style="color: #005050;">&#124;</FONT>&nbsp;
 <a href="http://genome.imim.es/main/research.html"
  onMouseover="window.status='Research at our Group';">Research</a>
&nbsp;<FONT style="background: #005050; color: #FFFFFF;">&nbsp;
 <a href="http://genome.imim.es/main/software.html"
  style="background: #005050; color: #DDDDDD;"
  onMouseover="window.status='Software developed in our Group';">Software</a>
&nbsp;</FONT>&nbsp; <!-- Current menu -->
 <a href="http://genome.imim.es/main/publications.html"
  onMouseover="window.status='Publications by our Group';">Publications</a>
&nbsp;<FONT style="color: #005050;">&#124;</FONT>&nbsp;
 <a href="http://genome.imim.es/main/links.html"
  onMouseover="window.status='Interesting Links';">Links</a>
</td>
</tr>
<tr>
<td class="links" align="CENTER">
 <a href="http://genome.imim.es/main/databases.html"
  onMouseover="window.status='Datasets created/curated by our Group';" class="hh">Resources &amp; Datasets</a>
&nbsp;<FONT style="color: #005050;">&#124;</FONT>&nbsp;
 <a href="http://genome.imim.es/genepredictions/index.html"
  onMouseover="window.status='Gene Predictions produced by our Software';" class="hh">Gene Predictions</a>
&nbsp;<FONT style="color: #005050;">&#124;</FONT>&nbsp;
 <a href="http://genome.imim.es/main/seminars.html"
  onMouseover="window.status='Seminars, Courses and Group Sessions';" class="hh">Seminars &amp; Courses</a>
</td>
</tr>
</table><BR class="hhh">

<!-- ############### PATH AREA ############### -->
<table border=1 cellpadding=0 cellspacing=0 width="100%">
<tr>
<td class="path" border=0 cellpadding=0 cellspacing=0 ALIGN="LEFT">
 <table border=0 cellpadding=0 cellspacing=1>
 <tr>
 <td class="ppath" border=0 cellpadding=0 cellspacing=0>
&nbsp;
<a href="http://www.imim.es/"
 onMouseover="window.status='Institut Municipal d\'Investigaci&oacute; M&egrave;dica';">IMIM</a>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<IMG SRC="http://genome.imim.es/g_icons/dmd.gif" ALT="*" HEIGHT=8 WIDTH=8 BORDER=0>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<a href="http://www.upf.es/"
 onMouseover="window.status='Universitat Pompeu Fabra';">UPF</a>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<IMG SRC="http://genome.imim.es/g_icons/dmd.gif" ALT="*" HEIGHT=8 WIDTH=8 BORDER=0>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<a href="http://www.crg.es/"
 onMouseover="window.status='Centre de Regulaci&oacute; Gen&ograve;mica';">CRG</a>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<IMG SRC="http://genome.imim.es/g_icons/dmd.gif" ALT="*" HEIGHT=8 WIDTH=8 BORDER=0>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<a href="http://www.imim.es/grib/eng/default.htm"
 onMouseover="window.status='Grup de Recerca en Informàtica Biom&egrave;dica';">GRIB</a>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<A HREF="http://genome.imim.es/index.html" onMouseover="window.status='Genome Informatics Research Laboratory HOME PAGE: You are Welcome!!!'">
<IMG SRC="http://genome.imim.es/g_icons/lhr.gif" ALT="HOME" HEIGHT=10 WIDTH=36 BORDER=0>
</A>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<a href="http://genome.imim.es/main/software.html"
  onMouseover="window.status='Software';">SOFTWARE</a>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<IMG SRC="http://genome.imim.es/g_icons/dmd.gif" ALT="*" HEIGHT=8 WIDTH=8 BORDER=0>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<a href="http://genome.imim.es/software/gfftools/GFF2APLOT.html#TOP"
 onMouseover="window.status='This is the main page for GFF2APLOT.';">gff2aplot</a>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<IMG SRC="http://genome.imim.es/g_icons/dmd.gif" ALT="*" HEIGHT=8 WIDTH=8 BORDER=0>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<a href="http://genome.imim.es/software/gfftools/GFF2APLOT.html#HOWTOS"
 onMouseover="window.status='GFF2APLOT Tutorials...';">HowTo</a>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<IMG SRC="http://genome.imim.es/g_icons/dmd.gif" ALT="*" HEIGHT=8 WIDTH=8 BORDER=0>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<a href="http://genome.imim.es/software/gfftools/gff2aplot_docs/mhcregion/index.html#TOP"
 onMouseover="window.status='An Alignment from WU-BLAST';">SIM Data</a>
 </td>
 </tr>
 </table>
</td>
</tr>
</table>

<!-- ############### USERS AREA ############### -->
<table border=0 cellpadding=0 cellspacing=0 width="100%">
<tr border=0 cellpadding=0 cellspacing=0>
<td class="section" border=0 cellpadding=0 cellspacing=0 width="10px">&nbsp;</td>
<td class="section" ALIGN=CENTER>
 <CENTER><FONT size=6 class="tgen"> SIM Alignment on a Multigenic Syntenic Region </FONT></CENTER>
</td>
<td class="section" border=0 cellpadding=0 cellspacing=0 width="10px">&nbsp;</td>
</tr>
<tr border=0 cellpadding=0 cellspacing=0>
<td class="userspace" border=0 cellpadding=0 cellspacing=0 width="10px">&nbsp;</td>
<td class="userspace" border=0 cellpadding=0 cellspacing=0>

&nbsp;<BR>
<A NAME="TOC"></A>
<P>
<FONT size=5 class="K"> Summary </FONT> <BR class="kk"><BR>
We will visualize the output produced by SIM on the genomic MHC syntenic region between human and mouse. We will use a filter to convert its output into GFF, the C program <code>ali2gff</code> is included in the <code>gff2aplot</code> distribution. <BR>
Once we have generated the whole genomic region figure, we will zoom into a single gene annotation. This will illustrate how you can expand a given area from your pictures, to highlight single features on multigenic annotation plots for instance. <BR class="kk"><BR>
<B>NOTE.-</B> For the sake of clarity, we are going to use long names for the comand-line switches. See the command-line help if you prefer short names for those cases in which a short name is available. <BR>
Bitmaps for the examples were generated as PNGs (Portable Network Graphics). If your browser is not ready for such format yet, you can visualize the PDF or PS versions by clicking on the links below each snapshot. Links to customization files, log files, GFF input files and output PostScript figures, are also available on each command-line shown. <BR>
</P>
<P>
<FONT size=5 class="K"> Contents </FONT>
<UL>
  <LI><A HREF="index.html#ALIGFF" onMouseover="window.status='GFF from SIM';">Translating SIM output into GFF.</A> <BR></LI>
  <LI><A HREF="index.html#RAW" onMouseover="window.status='Raw output from GFF2APLOT';">Raw <code>gff2aplot</code> output.</A> <BR></LI>
  <LI><A HREF="index.html#CUST" onMouseover="window.status='Whole genomic MHC region plot';">Finishing whole genomic region plot.</A> <BR></LI>
  <LI><A HREF="index.html#ZOOM" onMouseover="window.status='Zooming into LMP2 gene region';">Zooming into single gene regions.</A> <BR></LI>
</UL> <BR class="hh">
</P>

<TABLE border=0 cellpadding=0 cellspacing=0 width="100%">
<TR>
<TD class="section" width=80% ALIGN=LEFT>
<a name="ALIGFF"></a><FONT size=5 class="K"> Parsing SIM Output </FONT>
</TD>
<TD class="section" width=20% ALIGN=RIGHT>
<a href="index.html#TOP"
 onMouseover="window.status='TOP of the PAGE';">
<IMG class="pnt" SRC="http://genome.imim.es/g_icons/top.gif" HEIGHT=15 WIDTH=15 BORDER=0></a>
</TD></TR>
</TABLE>
<P> <BR class="hh">
For illustrating this example we are going to visualize the alignment produced by the SIM program (<a href="index.html#huang:1991">Huang and Miller, 1991</a>) on the syntenic MHCII genomic regions between human (<a href="hs.fa">hs.fa</a>) and mouse (<a href="mm.fa">mm.fa</a>). The sequences being analized contain four human genes (LMP2, TAP1, LMP7 and TAP2) and their counterpart ones in mouse (<I>lmp</I>2, <I>tap</I>1, <I>lmp</I>7 and <I>tap</I>2). The following is an excerpt from that program output, when applied to our data set: <BR>
</P>
<P><CENTER><TABLE class="userspace" border=2 cellpadding=0 cellspacing=0>
<TR><TD class="ytbl">&nbsp;&nbsp;</TD><TD class="ytbl" ALIGN=LEFT>&nbsp;<BR class="hh"><B><PRE class="codefg" style="font-size: 10px;">
d {
  "sim96 <a href="hs.fa">hs.fa</a> <a href="mm.fa">mm.fa</a>
   M = 10, I = -10, V = -10, O = 60, E = 2"
}
s {
  <a href="hs.fa">"hs.fa"</a> 1 32000
  <a href="mm.fa">"mm.fa"</a> 1 35000
}
h {
   ">X87344"
   ">AF100956+AF027865"
}
a {
  s 18340
  b 863 804
  e 19762 20115
  l 863 804 890 831 79
  &#183;&#183;&#183; 
  l 19715 20068 19762 20115 85
}
  &#183;&#183;&#183; 
a {
  s 208
  b 8426 33675
  e 8541 33734
  l 8426 33675 8449 33698 92
  l 8506 33699 8541 33734 75
} <BR class="hh">
</PRE></B></TD><TD class="ytbl">&nbsp;&nbsp;</TD></TR>
</TABLE></CENTER></P>
<P>
Header fields contain information regarding program execution parameters, input fasta files provided and also the length of those sequences. Then you can find all the alignments and the locally aligned fragments that define each alignment. <BR>
That file can be translated into GFF records by using <code>ali2gff</code>, a C program which is also included in the <code>gff2aplot</code> distribution tarball. Here is an example of such a command-line: <BR>
</P>
<P><CENTER><TABLE class="userspace" border=0 cellpadding=0 cellspacing=0><TR><TD class="codebg">&nbsp;&nbsp;</TD><TD class="codebg" ALIGN=LEFT>&nbsp; <BR class="hh"><PRE class="codefg">
ali2gff -x "X87344"             \
        -y "AF100956+AF027865"  \
           <a href="hs-mm.sim">hs-mm.sim</a>            \
         > <a href="hs-mm.sim.gff">hs-mm.sim.gff</a> <BR class="hh">
</PRE></TD><TD class="codebg">&nbsp;&nbsp;</TD></TR></TABLE></CENTER></P>
<P>
Roughly speaking, "<code>s</code>" field defines the contents of the "<code>seqbounds</code>" GFF record. "<code>s</code>", "<code>b</code>" and "<code>e</code>"  subfields within each alignment set ("<code>a</code>") define the contents of the "<code>alignment</code>" GFF records, while the "<code>l</code>" subfield corresponds to each "<code>fragment</code>" GFF records. <BR>
</P>
<P><CENTER><TABLE class="userspace" border=2 cellpadding=0 cellspacing=0>
<TR><TD class="ytbl">&nbsp;&nbsp;</TD><TD class="ytbl" ALIGN=LEFT>&nbsp;<BR class="hh"><B><PRE class="codefg" style="font-size: 10px;">
X87344:AF100956+AF027865 sim96 seqbounds    1:1     32000:35000      .     .   .
X87344:AF100956+AF027865 sim96 alignment  863:804   19762:20115 18340.000 +:+ .:.
X87344:AF100956+AF027865 sim96 fragment   863:804     890:831       0.790 +:+ .:. ## cumulative nmp: 0.001
  &#183;&#183;&#183; 
X87344:AF100956+AF027865 sim96 fragment 19715:20068 19762:20115     0.850 +:+ .:. ## cumulative nmp: 0.562
  &#183;&#183;&#183; 
X87344:AF100956+AF027865 sim96 alignment 8426:33675  8541:33734   208.000 +:+ .:.
X87344:AF100956+AF027865 sim96 fragment  8426:33675  8449:33698     0.920 +:+ .:. ## cumulative nmp: 0.251
X87344:AF100956+AF027865 sim96 fragment  8506:33699  8541:33734     0.750 +:+ .:. ## cumulative nmp: 0.558 <BR class="hh">
</PRE></B></TD><TD class="ytbl">&nbsp;&nbsp;</TD></TR>
</TABLE><CENTER></P>
<P>
On the other hand, we already have the axes annotation records from GenBank, we filtered them and saved into two GFF files, one for human (<a href="hs.gff">hs.gff</a>) and one for mouse (<a href="mm.gff">mm.gff</a>). <BR>
</P>
<center><HR width="80%" class="rtbl"></center>
<P>
<A NAME="huang:1991"><B>X. Huang and W. Miller</B></A> <BR class="kk">
&nbsp;&nbsp;&nbsp;&nbsp;"A time-efficient, linear-space local similarity algorithm." <BR class="kk">
&nbsp;&nbsp;&nbsp;&nbsp;<I>Adv. Appl. Math.</I>, 12:337-357, 1991. <BR class="kk">
</P>


<TABLE border=0 cellpadding=0 cellspacing=0 width="100%">
<TR>
<TD class="section" width=80% ALIGN=LEFT>
<a name="RAW"></a><FONT size=5 class="K"> Raw Output </FONT>
</TD>
<TD class="section" width=20% ALIGN=RIGHT>
<a href="index.html#TOP"
 onMouseover="window.status='TOP of the PAGE';">
<IMG class="pnt" SRC="http://genome.imim.es/g_icons/top.gif" HEIGHT=15 WIDTH=15 BORDER=0></a>
</TD></TR>
</TABLE>
<P> <BR class="hh">
Putting together the human and mouse annotations with the alignment is an easy step. We are running <code>gff2aplot</code> with the default settings, just adding few labels to the figure via command-line switches ("<code>--title</code>", "<code>--subtitle</code>" and "<code>--percent-box-label</code>"). <BR>
</P>
<P><CENTER><TABLE class="userspace" border=0 cellpadding=0 cellspacing=0><TR><TD class="userspace" ALIGN=CENTER VALIGN=MIDDLE><CENTER><a href="hs-mm.sim.png"><IMG SRC="hs-mm.sim.s.png" ALT="Default SIM plot" HEIGHT=150 WIDTH=120 BORDER=1></a> <BR> &nbsp;[<a href="hs-mm.sim.png">PNG</a>]&nbsp;[<a href="hs-mm.sim.ps">PS</a>]&nbsp;[<a href="hs-mm.sim.pdf">PDF</a>]&nbsp; <BR>
</CENTER></TD><TD class="codebg">&nbsp;&nbsp;</TD><TD class="codebg" ALIGN=LEFT>&nbsp; <BR class="hh"><PRE class="codefg">
gff2aplot.pl                           \
    --verbose                          \
    --percent-box-label "Identity%"    \
    --title "Human/Mouse MHCII Region" \
    --subtitle ""                      \
    -- <a href="hs.gff">hs.gff</a>                          \
       <a href="mm.gff">mm.gff</a>                          \
       <a href="hs-mm.sim.gff">hs-mm.sim.gff</a>                   \
     > <a href="hs-mm.sim.ps">hs-mm.sim.ps</a>                    \
    2> <a href="hs-mm.sim.log">hs-mm.sim.log</a> <BR class="hh">
</PRE></TD><TD class="codebg">&nbsp;&nbsp;</TD></TR></TABLE></CENTER></P>
<P>
It is an extrange plot, isn't it ? We got an alignment diagonal which is an artifact product of the GFF input file we have used, although that file is well formated. It follows the APLOT pseudo-GFF format instead of the standard GFF one, but that must not alter the resulting plot. <br>
Perhaps you already found what is wrong with that picture. Yes, we have included as an alignment the "<code>seqbounds</code>", the <code>gff2aplot</code> does not have a "fixed" definition of alignment records. The program takes all the records for two sequences that are in APLOT format as they code for alignment data to be drawn in the alignment panel. The same occurs with the GFF-version2 records having a pre-defined grouping tag, '<code>target "group-id"</code>' (unless you have changed the alignment tag string via "<code>align_tag</code>" layout customization variable). <br>
We are going to set different colors to the three GFF-features found in the "<code>hs-mm.sim.gff</code>" file ("<code>seqbounds</code>", "<code>alignment</code>" and "<code>fragment</code>"). We are also going to force that all those records without a defined strand, as "<code>seqbounds</code>" in this example, will be drawn in the bottomest layer. In the next plot we are setting both parameters from command-line (the "<code>--feature-var</code>" and "<code>--strand-var</code>" switches, respectively). We are switching on the percent bottom panel too (with "<code>--show-percent-box</code>"). <br>
</P>
<P><CENTER><TABLE class="userspace" border=0 cellpadding=0 cellspacing=0><TR><TD class="userspace" ALIGN=CENTER VALIGN=MIDDLE><CENTER><a href="hs-mm.sim2.png"><IMG SRC="hs-mm.sim2.s.png" ALT="Command-line Customization" HEIGHT=150 WIDTH=120 BORDER=1></a> <BR> &nbsp;[<a href="hs-mm.sim2.png">PNG</a>]&nbsp;[<a href="hs-mm.sim2.ps">PS</a>]&nbsp;[<a href="hs-mm.sim2.pdf">PDF</a>]&nbsp; <BR>
</CENTER></TD><TD class="codebg">&nbsp;&nbsp;</TD><TD class="codebg" ALIGN=LEFT>&nbsp; <BR class="hh"><PRE class="codefg">
gff2aplot.pl                                         \
     --verbose                                       \
     --percent-box-label "Identity%"                 \
     --show-percent-box                              \
     --title "Human/Mouse MHCII Region"              \
     --subtitle ""                                   \
     --feature-var "seqbounds::alignment_color=grey" \
     --feature-var "alignment::alignment_color=red"  \
     --feature-var "fragment::alignment_color=green" \
     --strand-var "..::strand_layer=0"               \
     --strand-var "++::strand_layer=2"               \
     -- <a href="hs.gff">hs.gff</a>                                       \
        <a href="mm.gff">mm.gff</a>                                       \
        <a href="hs-mm.sim.gff">hs-mm.sim.gff</a>                                \
      > <a href="hs-mm.sim2.ps">hs-mm.sim2.ps</a>                                \
     2> <a href="hs-mm.sim2.log">hs-mm.sim2.log</a> <BR class="hh">
</PRE></TD><TD class="codebg">&nbsp;&nbsp;</TD></TR></TABLE></CENTER></P>
<P>
</P>


<TABLE border=0 cellpadding=0 cellspacing=0 width="100%">
<TR>
<TD class="section" width=80% ALIGN=LEFT>
<a name="CUST"></a><FONT size=5 class="K"> Finishing Plot for MHC Region </FONT>
</TD>
<TD class="section" width=20% ALIGN=RIGHT>
<a href="index.html#TOP"
 onMouseover="window.status='TOP of the PAGE';">
<IMG class="pnt" SRC="http://genome.imim.es/g_icons/top.gif" HEIGHT=15 WIDTH=15 BORDER=0></a>
</TD></TR>
</TABLE>
<P> <BR class="hh">
To make the last figure we have changed few customization options to highlight an artifact, the "<code>seqbounds</code>" GFF record on the input file. We can remove those fields that are not related with the alignment after generating the GFF file, or we can hide those features that are not relevant for the final picture. We do not want to show the "<code>seqbounds</code>" records although they may be useful for other programs using the same GFF file, maybe in such cases we do not want to remove those records from the file. There is an alternative solution, just hide all the records containing such features. The following picture only shows the alignment fragments. We have defined that "<code>seqbounds</code>" and "<code>alignment</code>" features must be hidden, in this case by setting the correpsonding variable in the customization file (see <a href="mhc2.rc">mhc2.rc</a>).
</P>
<P><CENTER><TABLE class="userspace" border=0 cellpadding=0 cellspacing=0><TR><TD class="userspace" ALIGN=CENTER VALIGN=MIDDLE><CENTER><a href="hs-mm.sim.ini.png"><IMG SRC="hs-mm.sim.ini.s.png" ALT="Empty Plot" HEIGHT=150 WIDTH=120 BORDER=1></a> <BR> &nbsp;[<a href="hs-mm.sim.ini.png">PNG</a>]&nbsp;[<a href="hs-mm.sim.ini.ps">PS</a>]&nbsp;[<a href="hs-mm.sim.ini.pdf">PDF</a>]&nbsp; <BR>
</CENTER></TD><TD class="codebg">&nbsp;&nbsp;</TD><TD class="codebg" ALIGN=LEFT>&nbsp; <BR class="hh"><PRE class="codefg">
gff2aplot.pl                            \
    --verbose                           \
    --percent-box-label "Identity%"     \
    --show-percent-box                  \
    --title  "Human/Mouse MHCII Region" \
    --subtitle ""                       \
    --x-label "Human" --y-label "Mouse" \
    --custom-filename <a href="mhc2.rc">mhc2.rc</a>           \
    --custom-filename <a href="mhc2spc.rc">mhc2spc.rc</a>        \
    -- <a href="hs.gff">hs.gff</a>                           \
       <a href="mm.gff">mm.gff</a>                           \
       <a href="hs-mm.sim.gff">hs-mm.sim.gff</a>                    \
     > <a href="hs-mm.sim.ini.ps">hs-mm.sim.ini.ps</a>                 \
    2> <a href="hs-mm.sim.ini.log">hs-mm.sim.ini.log</a> <BR class="hh">
</PRE></TD><TD class="codebg">&nbsp;&nbsp;</TD></TR></TABLE></CENTER></P>
<P>
We have said that we were going to hide the "<code>alignment</code>" GFF-features from the plot and we just changed its color. What's happening here ? <BR>
We simply provided an extra customization file (see <a href="mhc2.rc">mhc2spc.rc</a>). That file contains two variable settings which override the ones defined in the previous customization file (<a href="mhc2.rc">mhc2.rc</a>). It is a must to notice here that the order of the customization files (as for the GFF input records) is very important. We do not get the same results if we swap the command-line switches and we provide "<a href="mhc2.rc">mhc2spc.rc</a>" before "<a href="mhc2.rc">mhc2.rc</a>". The example also illustrates the advantatges of having a main customization file upon which we can change paramters of the figure (mainly via command-line switches, but via extra customization files too).
</P>


<TABLE border=0 cellpadding=0 cellspacing=0 width="100%">
<TR>
<TD class="section" width=80% ALIGN=LEFT>
<a name="ZOOM"></a><FONT size=5 class="K"> Zooming Into the LMP2 Gene Region </FONT>
</TD>
<TD class="section" width=20% ALIGN=RIGHT>
<a href="index.html#TOP"
 onMouseover="window.status='TOP of the PAGE';">
<IMG class="pnt" SRC="http://genome.imim.es/g_icons/top.gif" HEIGHT=15 WIDTH=15 BORDER=0></a>
</TD></TR>
</TABLE>
<P> <BR class="hh">
In this tutorial we would like to introduce the "zoom" feature too. When working with large datasets we can take two approaches: making larger plots (I mean using larger paper sizes) or zooming into interesting plot regions. There are two command-line switches that make that option possible without modifying the GFF input records (for instance, by filtering those records outside the zoomed region boundaries). "<code>--zoom-area</code>" allows user to show where is the area being expanded and "<code>--zoom</code>" which performs the real zoom into that area. Both can take up to four optional parameters, set from command-line or customization files, that define the X and Y nucleotide coords of the enlarged plot region. In this case, we are assuming that the zoom area will start on both sequences origins, otherwise you can define that with "<code>--start-x-sequence</code>" and "<code>--start-y-sequence</code>" command-line options. "<code>--end-x-sequence</code>" and "<code>--end-y-sequence</code>" define the final coords for the zoomed area. <BR>
</P>
<P><CENTER><TABLE class="userspace" border=0 cellpadding=0 cellspacing=0><TR><TD class="userspace" ALIGN=CENTER VALIGN=MIDDLE><CENTER><a href="hs-mm.sim.zbox.png"><IMG SRC="hs-mm.sim.zbox.s.png" ALT="Marking area to be zoomed" HEIGHT=150 WIDTH=120 BORDER=1></a> <BR> &nbsp;[<a href="hs-mm.sim.zbox.png">PNG</a>]&nbsp;[<a href="hs-mm.sim.zbox.ps">PS</a>]&nbsp;[<a href="hs-mm.sim.zbox.pdf">PDF</a>]&nbsp; <BR>
</CENTER></TD><TD class="codebg">&nbsp;&nbsp;</TD><TD class="codebg" ALIGN=LEFT>&nbsp; <BR class="hh"><PRE class="codefg">
gff2aplot.pl                                       \
    --verbose                                      \
    --percent-box-label "Identity%"                \
    --show-percent-box                             \
    --title  "Human/Mouse MHCII Region"            \
    --subtitle "Zooming into the LMP2 gene region" \
    --zoom-area                                    \
    --end-x-sequence 6500                          \
    --end-y-sequence 6500                          \
    --x-label "Human X87344"                       \
    --y-label "Mouse AF100956+AF027865"            \
    --custom-filename <a href="mhc2.rc">mhc2.rc</a>                      \
    --custom-filename <a href="mhc2labels.rc">mhc2labels.rc</a>                \
    -- <a href="hs.gff">hs.gff</a>                                      \
       <a href="mm.gff">mm.gff</a>                                      \
       <a href="hs-mm.sim.gff">hs-mm.sim.gff</a>                               \
     > <a href="hs-mm.sim.zbox.ps">hs-mm.sim.zbox.ps</a>                           \
    2> <a href="hs-mm.sim.zbox.log">hs-mm.sim.zbox.log</a> <BR class="hh">
</PRE></TD><TD class="codebg">&nbsp;&nbsp;</TD></TR></TABLE></CENTER></P>
<P>
Above picture highlights the area being enlarged in the plot below. We defined the fill color of the highlighted area (set to violet in <a href="mhc2.rc">mhc2.rc</a>) to be the same for the bottom figure (set to violet in <a href="mhc2+sgp1.rc">mhc2+sgp1.rc</a>). We are including gene predictions obtained by SGP1 program (<a href="index.html#wiehe:2001">Wiehe et al, 2001</a>). In both plots, the group (say here "genes") labels angle is set on "<a href="mhc2labels.rc">mhc2labels.rc</a>" customization file.
</P>
<P><CENTER><TABLE class="userspace" border=0 cellpadding=0 cellspacing=0><TR><TD class="userspace" ALIGN=CENTER VALIGN=MIDDLE><CENTER><a href="hs-mm.sim.zoom.png"><IMG SRC="hs-mm.sim.zoom.s.png" ALT="Zooming into LMP2 region" HEIGHT=150 WIDTH=120 BORDER=1></a> <BR> &nbsp;[<a href="hs-mm.sim.zoom.png">PNG</a>]&nbsp;[<a href="hs-mm.sim.zoom.ps">PS</a>]&nbsp;[<a href="hs-mm.sim.zoom.pdf">PDF</a>]&nbsp; <BR>
</CENTER></TD><TD class="codebg">&nbsp;&nbsp;</TD><TD class="codebg" ALIGN=LEFT>&nbsp; <BR class="hh"><PRE class="codefg">
gff2aplot.pl                                       \
    --verbose                                      \
    --percent-box-label "Identity%"                \
    --show-percent-box                             \
    --title  "Human/Mouse MHCII Region"            \
    --subtitle "Zooming into the LMP2 gene region" \
    --zoom                                         \
    --end-x-sequence 6500                          \
    --end-y-sequence 6500                          \
    --x-label "Human"                              \
    --y-label "Mouse"                              \
    --custom-filename <a href="mhc2.rc">mhc2.rc</a>                      \
    --custom-filename <a href="mhc2labels.rc">mhc2labels.rc</a>                \
    --custom-filename <a href="mhc2+sgp1.rc">mhc2+sgp1.rc</a>                 \
    -- <a href="hs.gff">hs.gff</a>                                      \
       <a href="mm.gff">mm.gff</a>                                      \
       <a href="hs-mm.sgp1.gff">hs-mm.sgp1.gff</a>                              \
       <a href="hs-mm.sim.gff">hs-mm.sim.gff</a>                               \
     > <a href="hs-mm.sim.zoom.ps">hs-mm.sim.zoom.ps</a>                           \
    2> <a href="hs-mm.sim.zoom.log">hs-mm.sim.zoom.log</a> <BR class="hh">
</PRE></TD><TD class="codebg">&nbsp;&nbsp;</TD></TR></TABLE></CENTER></P>
<center><HR width="80%" class="rtbl"></center>
<P>
<A NAME="wiehe:2001"><B>T. Wiehe, S. Gebauer-Jung, T. Mitchell-Olds and R. Guig&oacute;</B></A> <BR class="kk">
&nbsp;&nbsp;&nbsp;&nbsp;"SGP-1: prediction and validation of homologous genes based on sequence alignments." <BR class="kk">
&nbsp;&nbsp;&nbsp;&nbsp;<I>Genome Research</I>, 11(9):1574-1583, 2001. <BR class="kk">
</P>

</td>
<td class="userspace" border=0 cellpadding=0 cellspacing=0 width="10px">&nbsp;</td>
</tr>
</table>

<!-- ############### TRAILER AREA ############### -->
<table class="trailer" border=0 cellpadding=0 cellspacing=0 width="100%">
 <tr class="default">
 <td id="traila" align=left>
&nbsp;
<A HREF="http://genome.imim.es/main/disclaimer.html"
 onMouseover="window.status='Data Release Policy and Software License';">Disclaimer</A>
 </td>
 <td id="trailb" align=center>
<SCRIPT type="text/javascript" languaje="javascript">
<!--
 LastUpdateSplit()
// -->
</SCRIPT>
 </td>
 <td id="trailc" align=rigth>
<A HREF="MAILTO:gmaster@imim.es"
 onMouseover="window.status='Contact with our webmaster';">webmaster</A>
&nbsp;
 </td>
 </tr>
</table>

<SCRIPT type="text/javascript" languaje="javascript">
<!--
 reportErrors();
// -->
</SCRIPT>
<!-- ############### EOF ############### -->
</body>
</html>
