<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
<head>
 <META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">
 <META NAME="Generator" CONTENT="Hacker Hands">
 <META NAME="author" CONTENT="Josep F. Abril">
  <!-- $Id: index.html,v 1.1 2003-06-10 10:33:28 jabril Exp $ -->
  <META NAME="keywords"    CONTENT="GFF2APLOT HOWTOs, tutorials, GFF, gff-format, Genomic Pairwise Alignments PostScript Plots, Genomic Analysis, Homology, Sinteny, Bioinformatics, comparative analysis of splice sites">
  <META NAME="description" CONTENT="GFF2APLOT HOWTOs: Splice sites comparative analysis.">
 <TITLE> GFF2APLOT HOWTOs: Splice sites comparative analysis. </TITLE>
<!-- <BASE HREF="http://genome.imim.es/software/gfftools/gff2aplot_docs/splice_sites/"> -->
<link rel=stylesheet type="text/css" href="http://genome.imim.es/Genome.css" title="Genome">
<link rel=stylesheet type="text/css" href="http://genome.imim.es/software/gfftools/gff2aplot_docs/HowTo.css" title="Genome">
<STYLE type="text/css">
 @import url(http://genome.imim.es/Genome_forms_mozilla.css);
</STYLE>
<SCRIPT type="text/javascript" src="http://genome.imim.es/Genome.js"></SCRIPT>
</head>

<body onResize="init();" MARGINWIDTH="2" MARGINHEIGHT="0">
<A NAME="TOP"></A>

<!-- ############### TITLE AREA ############### -->
<table border=0 cellpadding=0 cellspacing=0 width="100%">
<tr>
<td align="center">
<FONT size=4 class="tit">Genome BioInformatics Research Lab</FONT>
</td>
</tr>
</table>

<!-- ############### MENU AREA ############### -->
<table class="default" border=0 cellpadding=0 cellspacing=0 width="100%">
<tr>
<td class="links" align="CENTER">
 <a href="http://genome.imim.es/main/help.html" 
  onMouseover="window.status='Asking for Help';">Help</a>
&nbsp;<FONT style="color: #005050;">&#124;</FONT>&nbsp;
 <a href="http://genome.imim.es/main/news.html"
  onMouseover="window.status='News';">News</a>
&nbsp;<FONT style="color: #005050;">&#124;</FONT>&nbsp;
 <a href="http://genome.imim.es/main/people.html"
  onMouseover="window.status='Who is who in Genome Informatics Research Laboratory';">People</a>
&nbsp;<FONT style="color: #005050;">&#124;</FONT>&nbsp;
 <a href="http://genome.imim.es/main/research.html"
  onMouseover="window.status='Research at our Group';">Research</a>
&nbsp;<FONT style="background: #005050; color: #FFFFFF;">&nbsp;
 <a href="http://genome.imim.es/main/software.html"
  style="background: #005050; color: #DDDDDD;"
  onMouseover="window.status='Software developed in our Group';">Software</a>
&nbsp;</FONT>&nbsp; <!-- Current menu -->
 <a href="http://genome.imim.es/main/publications.html"
  onMouseover="window.status='Publications by our Group';">Publications</a>
&nbsp;<FONT style="color: #005050;">&#124;</FONT>&nbsp;
 <a href="http://genome.imim.es/main/links.html"
  onMouseover="window.status='Interesting Links';">Links</a>
</td>
</tr>
<tr>
<td class="links" align="CENTER">
 <a href="http://genome.imim.es/main/databases.html"
  onMouseover="window.status='Datasets created/curated by our Group';" class="hh">Resources &amp; Datasets</a>
&nbsp;<FONT style="color: #005050;">&#124;</FONT>&nbsp;
 <a href="http://genome.imim.es/genepredictions/index.html"
  onMouseover="window.status='Gene Predictions produced by our Software';" class="hh">Gene Predictions</a>
&nbsp;<FONT style="color: #005050;">&#124;</FONT>&nbsp;
 <a href="http://genome.imim.es/main/seminars.html"
  onMouseover="window.status='Seminars, Courses and Group Sessions';" class="hh">Seminars &amp; Courses</a>
</td>
</tr>
</table><BR class="hhh">

<!-- ############### PATH AREA ############### -->
<table border=1 cellpadding=0 cellspacing=0 width="100%">
<tr>
<td class="path" border=0 cellpadding=0 cellspacing=0 ALIGN="LEFT">
 <table border=0 cellpadding=0 cellspacing=1>
 <tr>
 <td class="ppath" border=0 cellpadding=0 cellspacing=0>
&nbsp;
<a href="http://www.imim.es/"
 onMouseover="window.status='Institut Municipal d\'Investigaci&oacute; M&egrave;dica';">IMIM</a>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<IMG SRC="http://genome.imim.es/g_icons/dmd.gif" ALT="*" HEIGHT=8 WIDTH=8 BORDER=0>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<a href="http://www.upf.es/"
 onMouseover="window.status='Universitat Pompeu Fabra';">UPF</a>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<IMG SRC="http://genome.imim.es/g_icons/dmd.gif" ALT="*" HEIGHT=8 WIDTH=8 BORDER=0>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<a href="http://www.crg.es/"
 onMouseover="window.status='Centre de Regulaci&oacute; Gen&ograve;mica';">CRG</a>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<IMG SRC="http://genome.imim.es/g_icons/dmd.gif" ALT="*" HEIGHT=8 WIDTH=8 BORDER=0>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<a href="http://www.imim.es/grib/eng/default.htm"
 onMouseover="window.status='Grup de Recerca en Informàtica Biom&egrave;dica';">GRIB</a>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<A HREF="http://genome.imim.es/index.html" onMouseover="window.status='Genome Informatics Research Laboratory HOME PAGE: You are Welcome!!!'">
<IMG SRC="http://genome.imim.es/g_icons/lhr.gif" ALT="HOME" HEIGHT=10 WIDTH=36 BORDER=0>
</A>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<a href="http://genome.imim.es/main/software.html"
  onMouseover="window.status='Software';">SOFTWARE</a>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<IMG SRC="http://genome.imim.es/g_icons/dmd.gif" ALT="*" HEIGHT=8 WIDTH=8 BORDER=0>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<a href="http://genome.imim.es/software/gfftools/GFF2APLOT.html#TOP"
 onMouseover="window.status='This is the main page for GFF2APLOT.';">gff2aplot</a>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<IMG SRC="http://genome.imim.es/g_icons/dmd.gif" ALT="*" HEIGHT=8 WIDTH=8 BORDER=0>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<a href="http://genome.imim.es/software/gfftools/GFF2APLOT.html#HOWTOS"
 onMouseover="window.status='GFF2APLOT Tutorials...';">HowTo</a>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<IMG SRC="http://genome.imim.es/g_icons/dmd.gif" ALT="*" HEIGHT=8 WIDTH=8 BORDER=0>
 </td><td class="ppath" border=0 cellpadding=0 cellspacing=0>
<a href="http://genome.imim.es/software/gfftools/gff2aplot_docs/splice_sites/index.html#TOP"
 onMouseover="window.status='Splice sites comparative analysis';">Splice Sites</a>
 </td>
 </tr>
 </table>
</td>
</tr>
</table>

<!-- ############### USERS AREA ############### -->
<table border=0 cellpadding=0 cellspacing=0 width="100%">
<tr border=0 cellpadding=0 cellspacing=0>
<td class="section" border=0 cellpadding=0 cellspacing=0 width="10px">&nbsp;</td>
<td class="section" ALIGN=CENTER>
 <CENTER><FONT size=6 class="tgen"> Comparative Analysis of Orthologous Splice Sites </FONT></CENTER>
</td>
<td class="section" border=0 cellpadding=0 cellspacing=0 width="10px">&nbsp;</td>
</tr>
<tr border=0 cellpadding=0 cellspacing=0>
<td class="userspace" border=0 cellpadding=0 cellspacing=0 width="10px">&nbsp;</td>
<td class="userspace" border=0 cellpadding=0 cellspacing=0>

&nbsp;<BR>
<A NAME="TOC"></A>
<P>
<FONT size=5 class="K"> Summary </FONT> <BR class="kk"><BR>
We will apply colors to different alignment features to visualize relationships among splice sites and the current gene annotation for an orthologous sequence pair between human and mouse. You will find this useful when integrating different data into a single figure. This tutorial also shows how to apply score-dependent color gradients to alignment features. You will learn how to project axes annotation features into the alignment panels, for instance, to pinpoint the real splice junctions from the rest. Current version of <code>gff2aplot</code> allows to emulate color blending to enhance the visualization of those projections. <BR class="kk"><BR>
<B>NOTE.-</B> For the sake of clarity, we are going to use long names for the comand-line switches. See the command-line help if you prefer short names for those cases in which a short name is available. <BR>
Bitmaps for the examples were generated as PNGs (Portable Network Graphics). If your browser is not ready for such format yet, you can visualize the PDF or PS versions by clicking on the links below each snapshot. Links to customization files, log files, GFF input files and output PostScript figures, are also available on each command-line shown. <BR>
</P>
<P>
<FONT size=5 class="K"> Contents </FONT>
<UL>
  <LI><A HREF="index.html#SEPAR" onMouseover="window.status='Donors or Acceptors';">Visualizing donor or acceptor homologous splice sites.</A> <BR></LI>
  <LI><A HREF="index.html#MERGE" onMouseover="window.status='Merging Donors and Acceptors';">Merging donor and acceptor sites plots.</A> <BR></LI>
  <LI><A HREF="index.html#CUST" onMouseover="window.status='Projecting axes annotations';">Projecting axes annotation on alignment panels.</A> <BR></LI>
  <LI><A HREF="index.html#BLEND" onMouseover="window.status='Pseudo colors blend on projections';">Color blending on axes annotation projections.</A> <BR></LI>
</UL> <BR class="hh">
</P>

<TABLE border=0 cellpadding=0 cellspacing=0 width="100%">
<TR>
<TD class="section" width=80% ALIGN=LEFT>
<a name="SEPAR"></a><FONT size=5 class="K"> Visualizing Orthologous Splice Sites </FONT>
</TD>
<TD class="section" width=20% ALIGN=RIGHT>
<a href="index.html#TOP"
 onMouseover="window.status='TOP of the PAGE';">
<IMG class="pnt" SRC="http://genome.imim.es/g_icons/top.gif" HEIGHT=15 WIDTH=15 BORDER=0></a>
</TD></TR>
</TABLE>
<P> <BR class="hh">
All subsequences matching GT (splicing donors) and AG (splicing acceptors) were extracted from the KCRB locus human and mouse orthologous sequences. A gene prediction tool was used to score their likelyhood to be real splice sites, based on position weight matrices specific for donor or acceptor sites. After that, homologous sites were encoded into APLOT pseudo-GFF format, saved as "<a href="KCRB.GT.hs-mm.gff">KCRB.GT.hs-mm.gff</a>" and "<a href="KCRB.AG.hs-mm.gff">KCRB.AG.hs-mm.gff</a>". <BR>
</P>
<P><CENTER><TABLE class="userspace" border=0 cellpadding=0 cellspacing=0><TR><TD class="userspace" ALIGN=CENTER VALIGN=MIDDLE><CENTER><a href="KCRB.GT.hs-mm.png"><IMG SRC="KCRB.GT.hs-mm.s.png" ALT="Donors Plot" HEIGHT=150 WIDTH=120 BORDER=1></a> <BR> &nbsp;[<a href="KCRB.GT.hs-mm.png">PNG</a>]&nbsp;[<a href="KCRB.GT.hs-mm.ps">PS</a>]&nbsp;[<a href="KCRB.GT.hs-mm.pdf">PDF</a>]&nbsp; <BR>
</CENTER></TD><TD class="codebg">&nbsp;&nbsp;</TD><TD class="codebg" ALIGN=LEFT>&nbsp; <BR class="hh"><PRE class="codefg">
gff2aplot.pl              \
    --verbose             \
    --title "DONOR SITES" \
    -- <a href="KCRB.hs.gff">KCRB.hs.gff</a>        \
       <a href="KCRB.mm.gff">KCRB.mm.gff</a>        \
       <a href="KCRB.GT.hs-mm.gff">KCRB.GT.hs-mm.gff</a>  \
     > <a href="KCRB.GT.hs-mm.ps">KCRB.GT.hs-mm.ps</a>   \
    2> <a href="KCRB.GT.hs-mm.log">KCRB.GT.hs-mm.log</a> <BR class="hh">
</PRE></TD><TD class="codebg">&nbsp;&nbsp;</TD></TR></TABLE></CENTER></P>
<P>
Above figure shows the output of <code>gff2aplot</code> when it was run with the human/mouse donor sites analysis. Below one was made for the human/mouse acceptor sites. Both were performed under program default settings. <BR>
</P>
<P><CENTER><TABLE class="userspace" border=0 cellpadding=0 cellspacing=0><TR><TD class="userspace" ALIGN=CENTER VALIGN=MIDDLE><CENTER><a href="KCRB.AG.hs-mm.png"><IMG SRC="KCRB.AG.hs-mm.s.png" ALT="Acceptors Plot" HEIGHT=150 WIDTH=120 BORDER=1></a> <BR> &nbsp;[<a href="KCRB.AG.hs-mm.png">PNG</a>]&nbsp;[<a href="KCRB.AG.hs-mm.ps">PS</a>]&nbsp;[<a href="KCRB.AG.hs-mm.pdf">PDF</a>]&nbsp; <BR>
</CENTER></TD><TD class="codebg">&nbsp;&nbsp;</TD><TD class="codebg" ALIGN=LEFT>&nbsp; <BR class="hh"><PRE class="codefg">
gff2aplot.pl                 \
    --verbose                \
    --title "ACCEPTOR SITES" \
    -- <a href="KCRB.hs.gff">KCRB.hs.gff</a>           \
       <a href="KCRB.mm.gff">KCRB.mm.gff</a>           \
       <a href="KCRB.AG.hs-mm.gff">KCRB.AG.hs-mm.gff</a>     \
     > <a href="KCRB.AG.hs-mm.ps">KCRB.AG.hs-mm.ps</a>      \
    2> <a href="KCRB.AG.hs-mm.log">KCRB.AG.hs-mm.log</a> <BR class="hh">
</PRE></TD><TD class="codebg">&nbsp;&nbsp;</TD></TR></TABLE></CENTER></P>
<P>
</P>


<TABLE border=0 cellpadding=0 cellspacing=0 width="100%">
<TR>
<TD class="section" width=80% ALIGN=LEFT>
<a name="MERGE"></a><FONT size=5 class="K"> Merging Alignment Plots </FONT>
</TD>
<TD class="section" width=20% ALIGN=RIGHT>
<a href="index.html#TOP"
 onMouseover="window.status='TOP of the PAGE';">
<IMG class="pnt" SRC="http://genome.imim.es/g_icons/top.gif" HEIGHT=15 WIDTH=15 BORDER=0></a>
</TD></TR>
</TABLE>
<P> <BR class="hh">
In this splice sites analysis could be interesting to plot together donors and acceptors. It is as easy as putting both human and mouse files into a single command-line, as follows: <BR>
</P>
<P><CENTER><TABLE class="userspace" border=0 cellpadding=0 cellspacing=0><TR><TD class="userspace" ALIGN=CENTER VALIGN=MIDDLE><CENTER><a href="KCRB.GT-AG.hs-mm.png"><IMG SRC="KCRB.GT-AG.hs-mm.s.png" ALT="Merging Donors and Acceptors" HEIGHT=150 WIDTH=120 BORDER=1></a> <BR> &nbsp;[<a href="KCRB.GT-AG.hs-mm.png">PNG</a>]&nbsp;[<a href="KCRB.GT-AG.hs-mm.ps">PS</a>]&nbsp;[<a href="KCRB.GT-AG.hs-mm.pdf">PDF</a>]&nbsp; <BR>
</CENTER></TD><TD class="codebg">&nbsp;&nbsp;</TD><TD class="codebg" ALIGN=LEFT>&nbsp; <BR class="hh"><PRE class="codefg">
gff2aplot.pl                                \
    --verbose                               \
    --title "DONOR + ACCEPTOR SPLICE SITES" \
    -- <a href="KCRB.hs.gff">KCRB.hs.gff</a>                          \
       <a href="KCRB.mm.gff">KCRB.mm.gff</a>                          \
       <a href="KCRB.GT.hs-mm.gff">KCRB.GT.hs-mm.gff</a>                    \
       <a href="KCRB.AG.hs-mm.gff">KCRB.AG.hs-mm.gff</a>                    \
     > <a href="KCRB.GT-AG.hs-mm.ps">KCRB.GT-AG.hs-mm.ps</a>                  \
    2> <a href="KCRB.GT-AG.hs-mm.log">KCRB.GT-AG.hs-mm.log</a> <BR class="hh">
</PRE></TD><TD class="codebg">&nbsp;&nbsp;</TD></TR></TABLE></CENTER></P>
<P>
The main drawback of this simple approach is that we cannot distinguish between donors and acceptors in the alignment panel. <BR>
</P>


<TABLE border=0 cellpadding=0 cellspacing=0 width="100%">
<TR>
<TD class="section" width=80% ALIGN=LEFT>
<a name="CUST"></a><FONT size=5 class="K"> Projecting Axes Annotation </FONT>
</TD>
<TD class="section" width=20% ALIGN=RIGHT>
<a href="index.html#TOP"
 onMouseover="window.status='TOP of the PAGE';">
<IMG class="pnt" SRC="http://genome.imim.es/g_icons/top.gif" HEIGHT=15 WIDTH=15 BORDER=0></a>
</TD></TR>
</TABLE>
<P> <BR class="hh">
In order to facilitate the analysis of the merged splice sites data set, we assign different colors to those features. We have named donor features as "<code>alignGT</code>" and acceptor ones as "<code>alignAG</code>", then we set the GFF-features variable "<code>alignment_color</code>" to darkred and darkblue respectively. We switch on the variable "<code>alignment_scale_color</code>" too, so feature score will define the color, the intensity of that color being directly proportional to the score. <BR>
We also want to take into account the known gene structures. It would be better if we could have a reference to the annotation coordinates. That is achieved by projecting such annotation features into the alignment panels. The variables controlling how to represent those projections are "<code>show_ribbons</code>", "<code>ribbon_style</code>" and "<code>ribbon_color</code>". As you can see on the customization file (take a look to "<a href="GT-AG.rc">GT-AG.rc</a>" file), we must switch on the projections, from now on "ribbons", and then set their color and/or style (if you like). We have choosen lightyellow and palegrey to fill exonic and UTR ribbons respectively. <BR>
</P>
<P><CENTER><TABLE class="userspace" border=0 cellpadding=0 cellspacing=0><TR><TD class="userspace" ALIGN=CENTER VALIGN=MIDDLE><CENTER><a href="KCRB.GT-AG.hs-mm.color.png"><IMG SRC="KCRB.GT-AG.hs-mm.color.s.png" ALT="Customizing" HEIGHT=150 WIDTH=120 BORDER=1></a> <BR> &nbsp;[<a href="KCRB.GT-AG.hs-mm.color.png">PNG</a>]&nbsp;[<a href="KCRB.GT-AG.hs-mm.color.ps">PS</a>]&nbsp;[<a href="KCRB.GT-AG.hs-mm.color.pdf">PDF</a>]&nbsp; <BR>
</CENTER></TD><TD class="codebg">&nbsp;&nbsp;</TD><TD class="codebg" ALIGN=LEFT>&nbsp; <BR class="hh"><PRE class="codefg">
$BIN/gff2aplot.pl                                        \
         --verbose                                       \
         --title "DONOR + ACCEPTOR SPLICE SITES"         \
         --subtitle                                      \
           "DONORs (GT) in RED - ACCEPTORs (AG) in BLUE" \
         --x-sequence-coords 0..4200                     \
         --y-sequence-coords 0..4600                     \
         --custom-filename <a href="GT-AG.rc">GT-AG.rc</a>                      \
         -- <a href="KCRB.hs.gff">KCRB.hs.gff</a>                                  \
            <a href="KCRB.mm.gff">KCRB.mm.gff</a>                                  \
            <a href="KCRB.GT.hs-mm.gff">KCRB.GT.hs-mm.gff</a>                            \
            <a href="KCRB.AG.hs-mm.gff">KCRB.AG.hs-mm.gff</a>                            \
          > <a href="KCRB.GT-AG.hs-mm.color.ps">KCRB.GT-AG.hs-mm.color.ps</a>                    \
         2> <a href="KCRB.GT-AG.hs-mm.color.log">KCRB.GT-AG.hs-mm.color.log</a> <BR class="hh">
</PRE></TD><TD class="codebg">&nbsp;&nbsp;</TD></TR></TABLE></CENTER></P>
<P>
You may agree that the current version is much more informative than any other previous one. <BR>
</P>


<TABLE border=0 cellpadding=0 cellspacing=0 width="100%">
<TR>
<TD class="section" width=80% ALIGN=LEFT>
<a name="BLEND"></a><FONT size=5 class="K"> Blending Projection Colors </FONT>
</TD>
<TD class="section" width=20% ALIGN=RIGHT>
<a href="index.html#TOP"
 onMouseover="window.status='TOP of the PAGE';">
<IMG class="pnt" SRC="http://genome.imim.es/g_icons/top.gif" HEIGHT=15 WIDTH=15 BORDER=0></a>
</TD></TR>
</TABLE>
<P> <BR class="hh">
The lastest plot is almost perfect altought the horizontal "ribbons" overlap all the vertical ones. PostScript does not allow transparent layers, but we can emulate then in a different ways. We have choosen complementary color-fill patterns to resolve the issue of the overlapping "ribbons". The parameters to enable this <code>gff2aplot</code> feature were collected into a new customization file: <a href="GT-AG_blend.rc">GT-AG_blend.rc</a>. The results can be seen below: <BR>
</P>
<P><CENTER><TABLE class="userspace" border=0 cellpadding=0 cellspacing=0><TR><TD class="userspace" ALIGN=CENTER VALIGN=MIDDLE><CENTER><a href="KCRB.GT-AG.hs-mm.color_blend.png"><IMG SRC="KCRB.GT-AG.hs-mm.color_blend.s.png" ALT="Color Pseudo-Blending" HEIGHT=150 WIDTH=120 BORDER=1></a> <BR> &nbsp;[<a href="KCRB.GT-AG.hs-mm.color_blend.png">PNG</a>]&nbsp;[<a href="KCRB.GT-AG.hs-mm.color_blend.ps">PS</a>]&nbsp;[<a href="KCRB.GT-AG.hs-mm.color_blend.pdf">PDF</a>]&nbsp; <BR>
</CENTER></TD><TD class="codebg">&nbsp;&nbsp;</TD><TD class="codebg" ALIGN=LEFT>&nbsp; <BR class="hh"><PRE class="codefg">
gff2aplot.pl                                        \
    --verbose                                       \
    --title "DONOR + ACCEPTOR SPLICE SITES (II)"    \
    --subtitle                                      \
      "DONORs (GT) in RED - ACCEPTORs (AG) in BLUE" \
    --x-sequence-coords 0..4200                     \
    --y-sequence-coords 0..4600                     \
    --custom-filename <a href="GT-AG.rc">GT-AG.rc</a>                      \
    --custom-filename <a href="GT-AG_blend.rc">GT-AG_blend.rc</a>                \
    -- <a href="KCRB.hs.gff">KCRB.hs.gff</a>                                  \
       <a href="KCRB.mm.gff">KCRB.mm.gff</a>                                  \
       <a href="KCRB.GT.hs-mm.gff">KCRB.GT.hs-mm.gff</a>                            \
       <a href="KCRB.AG.hs-mm.gff">KCRB.AG.hs-mm.gff</a>                            \
     > <a href="KCRB.GT-AG.hs-mm.color_blend.ps">KCRB.GT-AG.hs-mm.color_blend.ps</a>              \
    2> <a href="KCRB.GT-AG.hs-mm.color_blend.log">KCRB.GT-AG.hs-mm.color_blend.log</a> <BR class="hh">
</PRE></TD><TD class="codebg">&nbsp;&nbsp;</TD></TR></TABLE></CENTER></P>
<P>
The complementary pattern approach is device dependant, it only works on those devices implementing PostScript Level 2 or higher (current PostScript interpreter versions are releases of Level 3). It is also scale dependant, this means if you plan to reduce your plot, it is better to modify the "<code>color_merge_factor</code>". This variable allows numerical values between 0.07 and 2. That is what we have done on the new customization file provided, <a href="GT-AG_sm.rc">GT-AG_sm.rc</a>, increasing value from default 0.5 to 1. <BR>
</P>
<P><CENTER><TABLE class="userspace" border=0 cellpadding=0 cellspacing=0><TR><TD class="userspace" ALIGN=CENTER VALIGN=MIDDLE><CENTER><a href="KCRB.GT-AG.hs-mm.color_small.png"><IMG SRC="KCRB.GT-AG.hs-mm.color_small.s.png" ALT="Scaling Down Color Blend" HEIGHT=150 WIDTH=120 BORDER=1></a> <BR> &nbsp;[<a href="KCRB.GT-AG.hs-mm.color_small.png">PNG</a>]&nbsp;[<a href="KCRB.GT-AG.hs-mm.color_small.ps">PS</a>]&nbsp;[<a href="KCRB.GT-AG.hs-mm.color_small.pdf">PDF</a>]&nbsp; <BR>
</CENTER></TD><TD class="codebg">&nbsp;&nbsp;</TD><TD class="codebg" ALIGN=LEFT>&nbsp; <BR class="hh"><PRE class="codefg">
gff2aplot.pl                                        \
    --verbose                                       \
    --title "DONOR + ACCEPTOR SPLICE SITES (III)"   \
    --subtitle                                      \
      "DONORs (GT) in RED - ACCEPTORs (AG) in BLUE" \
    --x-sequence-coords 0..4200                     \
    --y-sequence-coords 0..4600                     \
    --custom-filename <a href="GT-AG.rc">GT-AG.rc</a>                      \
    --custom-filename <a href="GT-AG_sm.rc">GT-AG_sm.rc</a>                   \
    -- <a href="KCRB.hs.gff">KCRB.hs.gff</a>                                  \
       <a href="KCRB.mm.gff">KCRB.mm.gff</a>                                  \
       <a href="KCRB.GT.hs-mm.gff">KCRB.GT.hs-mm.gff</a>                            \
       <a href="KCRB.AG.hs-mm.gff">KCRB.AG.hs-mm.gff</a>                            \
     > <a href="KCRB.GT-AG.hs-mm.color_small.ps">KCRB.GT-AG.hs-mm.color_small.ps</a>              \
    2> <a href="KCRB.GT-AG.hs-mm.color_small.log">KCRB.GT-AG.hs-mm.color_small.log</a> <BR class="hh">
</PRE></TD><TD class="codebg">&nbsp;&nbsp;</TD></TR></TABLE></CENTER></P>
<P>
You can observe that now the "ribbons" look like pixelized. If you scale down twice the last figure, the overlapping regions for the annotation projections will print out like in the previous one (or pretty similar). <BR>
</P>


</td>
<td class="userspace" border=0 cellpadding=0 cellspacing=0 width="10px">&nbsp;</td>
</tr>
</table>

<!-- ############### TRAILER AREA ############### -->
<table class="trailer" border=0 cellpadding=0 cellspacing=0 width="100%">
 <tr class="default">
 <td id="traila" align=left>
&nbsp;
<A HREF="http://genome.imim.es/main/disclaimer.html"
 onMouseover="window.status='Data Release Policy and Software License';">Disclaimer</A>
 </td>
 <td id="trailb" align=center>
<SCRIPT type="text/javascript" languaje="javascript">
<!--
 LastUpdateSplit()
// -->
</SCRIPT>
 </td>
 <td id="trailc" align=rigth>
<A HREF="MAILTO:gmaster@imim.es"
 onMouseover="window.status='Contact with our webmaster';">webmaster</A>
&nbsp;
 </td>
 </tr>
</table>

<SCRIPT type="text/javascript" languaje="javascript">
<!--
 reportErrors();
// -->
</SCRIPT>
<!-- ############### EOF ############### -->
</body>
</html>
