<title>The DDBJ/EMBL/GenBank Feature Table Definition</title>

<h2>Feature key definitions</h2>

<pre>
<b>
<a name="allele">
Feature Key           allele</a>


Definition            a related individual or strain contains stable, alternative
                      forms of the same gene which differs from the presented
                      sequence at this location (and perhaps others)

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"
                      /frequency="text"
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /phenotype="text"
                      /product="text"
                      /standard_name="text"
                      /usedin=accnum:feature_label

Comment		      format of location is:
		      seq_location		/replace="variation"  
		      where variation is the sequence difference.  
                      Alternatively any valid location is acceptable and legal.


<hr>
<a name="attenuator">
Feature Key           attenuator</a>


Definition            1) region of DNA at which regulation of termination of
                         transcription occurs, which controls the expression
                         of some bacterial operons;
                      2) sequence segment located between the promoter and the
                         first structural gene that causes partial termination
                         of transcription

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"
                      /evidence=&lt;evidence_value&gt;
                      /label=feature_label
                      /gene="text"
                      /map="text"
                      /note="text"
                      /partial
                      /phenotype="text"
                      /usedin=accnum:feature_label

Organism scope        prokaryotes

Molecule scope        DNA

<hr>
<a name="c_region">
Feature Key           C_region</a>


Definition            Constant region of immunoglobulin light and heavy 
                      chains, and T-cell receptor alpha, beta, and gamma 
                      chains.  Includes one or more exons depending on the 
                      particular chain.

Optional qualifiers   /citation=[number]

                      /db_xref="&lt;database&gt;:&lt;identifier&gt"
                      /evidence=&lt;evidence_value&gt;
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /product="text"
                      /pseudo
                      /standard_name="text"
                      /usedin=accnum:feature_label



Parent Key            CDS

Organism scope        Eukaryotes

<hr>
<a name="caat_signal">
Feature Key           CAAT_signal</a>


Definition            CAAT box; part of a conserved sequence located about 75
                      bp up-stream of the start point of eukaryotic
                      transcription units which may be involved in RNA
                      polymerase binding; consensus=GG(C or T)CAATCT [1,2]

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /label=feature_label
                      /gene="text"
                      /map="text"
                      /note="text"
                      /partial
                      /usedin=accnum:feature_label

Organism scope        Eukaryotes and eukaryotic viruses

Molecule scope        DNA

References            [1]  Efstratiadis, A.  et al.  Cell 21, 653-668 (1980)
                      [2]  Nevins, J.R.  "The pathway of eukaryotic mRNA formation"  
                           Ann Rev Biochem 52, 441-466 (1983)


<hr>
<a name="cds">
Feature Key           CDS</a>


Definition            coding sequence; sequence of nucleotides that
                      corresponds with the sequence of amino acids in a
                      protein (location includes stop codon). 
                      Feature includes amino acid conceptual translation

Optional qualifiers   /citation=[number]
                      /codon=(seq:"text",aa:&lt;amino_acid&gt;)
                      /codon_start=&lt;1 or 2 or 3&gt;
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /EC_number="text"
                      /evidence=&lt;evidence_value&gt;
                      /exception="text"
                      /function="text"
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /number=&lt;integer&gt;
                      /partial
                      /product="text"
		      /protein_id="&lt;identifier&gt"
                      /pseudo
                      /standard_name="text"
                      /translation="text"
                      /transl_except=(pos:&lt;base_range&gt;,aa:&lt;amino_acid&gt;)
                      /transl_table =&lt;integer&gt;
                      /usedin=accnum:feature_label

Comment               /codon_start has valid value of 1 or 2 or 3, indicating
                      the offset at which the first complete codon of a coding
                      feature can be found, relative to the first base of
                      that feature.
                      /codon is used to specify unusual genetic codes,
                      including rare-usage start codons, organellar codes, etc.;
                      /codon is used to describe a single codon exception to
                      the code defined as the "normal" code for the organism;
                      it implies that the translation it specifies is used
                      throughout the feature
                      /codon specifies that a specific codon specified by
                      "seq" codes for the amino acid or stop codon specified
                      by "aa";
		      /protein_id onsists of a stable ID portion (3+5 format
		      with 3 position letters and 5 numbers) plus a version 
		      number after the decimal point.  When the protein 
		      sequence encoded by the CDS changes, only the version 
		      number of the /protein_id value is incremented. The
		      stable part of the /protein_id remains unchanged and as 
		      a result will permanently be associated with a given 
		      protein.
                      /transl_except is used to specify a single codon the
                      translation of which does not conform to the genetic
                      code defined by the "normal" code for organism or by the
                      /codon qualifiers given for the feature
                      /transl_table defines genetic code table used if other
                      than universal genetic code table. Genetic code
                      exception outside the range of the specified tables are
                      reported in /codon or /transl_except qualifiers

<hr>
<a name="conflict">
Feature Key           conflict</a>


Definition            independent determinations of the "same" sequence differ
                      at this site or region

Mandatory qualifiers  /citation=[number]

Optional qualifiers   /db_xref="&lt;database&gt;:&lt;identifier&gt"
		      /label=feature_label
		      /map="text"
                      /note="text"
                      /gene="text"
                      /usedin=accnum:feature_label

Comment               format of location is: 		
		      seq_location		/replace="conflict_location"  
		      where conflict_location  is the sequence difference.  
                      Alternatively any valid location is acceptable and legal.


<hr>
<a name="d-loop">
Feature Key           D-loop</a>


Definition            displacement loop; a region within mitochondrial DNA in
                      which a short stretch of RNA is paired with one strand
                      of DNA, displacing the original partner DNA strand in
                      this region; also used to describe the displacement of a
                      region of one strand of duplex DNA by a single stranded
                      invader in the reaction catalyzed by RecA protein

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                                  /evidence=&lt;evidence_value&gt;
                      /label=feature_label
                      /gene="text"
                      /map="text"
                      /note="text"
                      /partial
                      /usedin=accnum:feature_label

Molecule scope        DNA

<hr>
<a name="d_segment">
Feature Key           D_segment</a>


Definition            Diversity segment of immunoglobulin heavy chain, and 
                      T-cell receptor beta chain.  

Optional qualifiers   /citation=[number]
		      /db_xref="<database>:<identifier>"
		      /evidence=<evidence_value>
		      /gene="text"
		      /label=feature_label
		      /map="text"
		      /note="text"
 		      /partial
                      /product="text"
                      /pseudo
                      /standard_name="text"
                      /usedin=accnum:feature_label
                      
Parent Key            CDS

Organism scope        Eukaryotes

<hr>
<a name="enhancer">

Feature Key           enhancer


Definition            a cis-acting sequence that increases the utilization of
                      (some)  eukaryotic promoters, and can function in either
                      orientation and in any location (upstream or downstream)
                      relative to the promoter

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /label=feature_label
                      /gene="text"
                      /map="text"
                      /note="text"
                      /partial
                      /standard_name="text"
                      /usedin=accnum:feature_label

Organism scope        Eukaryotes and eukaryotic viruses

<hr>
<a name="exon">
Feature Key           exon</a>


Definition            region of genome that codes for portion of spliced mRNA;
                      may contain 5'UTR, all CDSs, and 3' UTR

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /EC_number="text"
                      /evidence=&lt;evidence_value&gt;
                      /function="text"
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /number=&lt;integer&gt;
                      /partial
                      /product="text"
                      /pseudo
                      /standard_name="text"
                      /usedin=accnum:feature_label


<hr>
<a name="gc_signal">
Feature Key           GC_signal</a>


Definition            GC box; a conserved GC-rich region located upstream of
                      the start point of eukaryotic transcription units which
                      may occur in multiple copies or in either orientation;
                      consensus=GGGCGG

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /usedin=accnum:feature_label

Organism scope        Eukaryotes and eukaryotic viruses

<hr>
<a name="gene">
Feature Key           gene</a>


Definition	      region of biological interest identified as a gene and for which a 
                      name has been assigned.

Optional qualifiers   /allele="text"
                      /citation=[number]
                      /db_xref="<database>:<identifier>"
                      /evidence=<evidence_value>
                      /function="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
		      /product="text"
		      /pseudo
                      /phenotype="text"
		      /standard_name="text"
                      /usedin=accnum:feature_label
	

Comment		      The gene feature describes  the interval of DNA that corresponds 
		      to a genetic trait or phenotype.  This feature is, by definition, 
	              not  strictly bound to it's positions at the ends.  It is meant to 
		      represent a region where the gene is located.
                                                             



<hr>
<a name="idna">
Feature Key           iDNA</a>


Definition            intervening DNA; DNA which is eliminated through any of
                      several kinds of recombination

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /function="text"
                      /label=feature_label
                      /gene="text"
                      /map="text"
                      /note="text"
                      /number=&lt;integer&gt;
                      /partial
                      /standard_name="text"
                      /usedin=accnum:feature_label

Molecule scope        DNA

Comment               e.g., in the somatic processing of immunoglobulin genes.

<hr>
<a name="intron">
Feature Key           intron</a>


Definition            a segment of DNA that is transcribed, but removed from
                      within the transcript by splicing together the sequences
                      (exons) on either side of it 

Optional qualifiers   /citation=[number]
                      /cons_splice=(5'site:&lt;bool&gt;,3'site:&lt;bool&gt;)
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /function="text"
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /number=&lt;integer&gt;
                      /partial
                      /standard_name="text"
                      /usedin=accnum:feature_label

Comment               cons_splice is used only when one of the intron's splice
                      sites does not match the GT...AG consensus.  Reference
                      for this consensus?

<hr>
<a name="j_segment">
Feature Key           J_segment</a>
 

Definition            Joining segment of immunoglobulin light and heavy 
                      chains, and T-cell receptor alpha, beta, and gamma 
                      chains.  

Optional qualifiers   /citation=[number]
		      /db_xref="&lt;database&gt;:&lt;identifier&gt;"
		      /evidence=&lt;evidence_value&gt;
		      /gene="text"

		      /map="text"
		      /note="text"
 		      /partial
                      /product="text"
                      /pseudo
                      /standard_name="text"
                      /usedin=accnum:feature_label

Parent Key            CDS

Organism scope        Eukaryotes

<hr>
<a name="ltr">
Feature Key           LTR</a>


Definition            long terminal repeat, a sequence directly repeated at
                      both ends of a defined sequence, of the sort typically
                      found in retroviruses

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /function="text"
                      /gene="text"
                      /label=feature_label
		      /map="text"
                      /note="text"
                      /partial
                      /standard_name="text"
                      /usedin=accnum:feature_label

<hr>
<a name="mat_peptide">
Feature Key           mat_peptide</a>


Definition            mature peptide or protein coding sequence; coding
                      sequence for the mature or final peptide or protein
                      product following post-translational modification.  The
                      location does not include the stop codon (unlike the
                      corresponding CDS)

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /EC_number="text"
                      /evidence=&lt;evidence_value&gt;
                      /function="text"
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /product="text"
		      /pseudo
                      /standard_name="text"
                      /usedin=accnum:feature_label


<hl>
<a name="misc_binding">
Feature Key           misc_binding</a>


Definition            site in nucleic acid which covalently or non-covalently
                      binds another moiety that cannot be described by any
                      other Binding key (primer_bind or protein_bind)

Mandatory qualifiers  /bound_moiety="text"

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /function="text"
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /usedin=accnum:feature_label

Comment               note that the key RBS is used for ribosome binding sites

<hr>
<a name="misc_difference">
Feature Key           misc_difference</a>


Definition            feature sequence is different from that presented in the
                      entry and cannot be described by any other Difference
                      key (conflict, unsure, old_sequence, mutation,
                      variation, allele, or modified_base)

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /clone="text"
                      /partial
                      /note="text"
                      /standard_name="text"
                      /usedin=accnum:feature_label

Comment               format of location is: 		
		      seq_location		/replace="change_location"  
	              where change_location is a sequence  descriptor or feature name that 
		      replaces the indicated sequence to yield the difference sequence.  
                      Alternatively any valid location is acceptable and legal.
<hr>
<a name="misc_feature">
Feature Key           misc_feature</a>


Definition            region of biological interest which cannot be described
                      by any other feature key; a new or rare feature

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /function="text"
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /number=&lt;integer&gt;
                      /phenotype="text"
                      /product="text"
                      /pseudo
                      /standard_name="text"
                      /usedin=accnum:feature_label

Comment               This key should not be used when the need is merely to mark a region in order to
                      comment on it or to use it in another feature's location; 
                      use the '-' pseudo-key instead.

<hr>
<a name="misc_recomb">
Feature Key           misc_recomb</a>


Definition            site of any generalized, site-specific or replicative
                      recombination event where there is a breakage and
                      reunion of duplex DNA that cannot be described by other
                      recombination keys (iDNA and virion) or qualifiers of
                      source key (/insertion_seq, /transposon, /proviral).

Mandatory qualifiers  /organism="text"

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /standard_name="text"
                      /usedin=accnum:feature_label

Molecule scope        DNA

<hr>
<a name="misc_rna">
Feature Key           misc_RNA</a>


Definition            any transcript or RNA product that cannot be defined by
                      other RNA keys (prim_transcript, precursor_RNA, mRNA,
                      5'clip, 3'clip, 5'UTR, 3'UTR, exon, CDS, sig_peptide,
                      transit_peptide, mat_peptide, intron, polyA_site, rRNA,
                      tRNA, scRNA, and snRNA)

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /function="text"
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /product="text"
                      /standard_name="text"
                      /usedin=accnum:feature_label

<hr>
<a name="misc_signal">
Feature Key           misc_signal</a>


Definition            any region containing a signal controlling or altering
                      gene function or expression that cannot be described by
                      other Signal keys (promoter, CAAT_signal, TATA_signal,
                      -35_signal, -10_signal, GC_signal, RBS, polyA_signal,
                      enhancer, attenuator, terminator, and rep_origin)

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /function="text"
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /phenotype="text"
                      /standard_name="text"
                      /usedin=accnum:feature_label

<hr>
<a name="misc_structure">
Feature Key           misc_structure</a>


Definition            any secondary or tertiary nucleotide structure or 
                      conformation that	cannot be described by other Structure
		      keys (stem_loop and D-loop)

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /function="text"
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /standard_name="text"
                      /usedin=accnum:feature_label

<hr>
<a name="modified_base">
Feature Key           modified_base</a>


Definition            the indicated nucleotide is a modified nucleotide and
                      should be substituted for by the indicated molecule
                      (given in the mod_base  qualifier value)

Mandatory qualifiers  /mod_base=&lt;modified_base&gt;

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /frequency="text"
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /usedin=accnum:feature_label

Comment               the mod_base value is limited to the restricted
                      vocabulary for modified base abbreviations

<hr>
<a name="mrna">
Feature Key           mRNA


Definition            messenger RNA; includes 5'untranslated region (5'UTR),
                      coding sequences (CDS, exon) and 3'untranslated region
                      (3'UTR)

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /exception="text"
                      /function="text"
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /product="text"
		      /pseudo
                      /standard_name="text"
                      /usedin=accnum:feature_label

<hr>
<a name="mutation">
Feature Key           mutation</a>


Definition            a related strain has an abrupt, inheritable change in
                      the sequence at this location

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /frequency="text"
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /phenotype="text"
                      /product="text"
                      /standard_name="text"
                      /usedin=accnum:feature_label

Comment               format of location is: 		
		      seq_location		/replace="mutation_location"  
		      where mutation_location  is the sequence difference and 
		      seq_location specifies the location in the presented
		      sequence o be replaced.
		      Alternatively any valid location is acceptable and legal.


<hr>
<a name="n_region">
Feature Key           N_region</a>


Definition            Extra nucleotides inserted between rearranged 
                      immmunoglobulin segments.

Optional qualifiers   /citation=[number]
		      /db_xref="<database>:<identifier>"
		      /evidence=<evidence_value>
		      /gene="text"
		      /label=feature_label
		      /map="text"
		      /note="text"
 		      /partial
                      /product="text"
                      /pseudo
                      /standard_name="text"
                      /usedin=accnum:feature_label

Parent Key            CDS

Organism scope        Eukaryotes

<hr>
<a name="old_sequence">
Feature Key	      old_sequence</a>


Definition            the presented sequence revises a previous version of the
                      sequence at this location

Mandatory qualifiers  /citation=[number]

Optional qualifiers   /db_xref="&lt;database&gt;:&lt;identifier&gt"
		      /gene="text"
		      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /usedin=accnum:feature_label

Comment               format of location is: 		
		      seq_location		/replace="change_loc"  
		      where change_loc represents the old sequence (before revision) and 
		      is usually a literal sequence and seq_location is the location in the 
		      presented (revised) sequence at which the revision took place.  To 
		      regenerate the sequence as it was before revision, replace the 
		      seq_location portion of the presented sequence with the sequence 
		      specified by change_loc.
                      Alternatively any valid location is acceptable and legal.


<hr>
<a name="polya_signal">
Feature Key           polyA_signal</a>


Definition            recognition region necessary for endonuclease cleavage
                      of an RNA transcript that is followed by polyadenylation;
                      consensus=AATAAA [1]

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /usedin=accnum:feature_label

Organism scope        Eukaryotes and eukaryotic viruses

References            [1] Proudfoot, N. and Brownlee, G.G. Nature 263, 211-214
                          (1976)

<hr>
<a name="polya_site">
Feature Key           polyA_site</a>


Definition            site on an RNA transcript to which will be added adenine
                      residues by post-transcriptional polyadenylation

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /usedin=accnum:feature_label

Organism scope        Eukaryotes and eukaryotic viruses

<hr>
<a name="precursor_rna">
Feature Key           precursor_RNA


Definition            any RNA species that is not yet the mature RNA product;
                      may include 5' clipped region (5'clip), 5' untranslated
                      region (5'UTR), coding sequences (CDS, exon),
                      intervening sequences (intron), 3' untranslated region
                      (3'UTR), and 3' clipped region (3'clip)

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"  
                      /evidence=&lt;evidence_value&gt;
                      /function="text"
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /standard_name="text"
                      /usedin=accnum:feature_label

Comment               this key is used for any RNA which may be the result of
                      some post-transcriptional processing.  If the RNA in
                      question is known not to have been processed, used the
                      prim_transcript key.

<hr>
<a name="prim_transcript">
Feature Key           prim_transcript</a>


Definition            primary (initial, unprocessed) transcript;  includes 5'
                      clipped region (5'clip), 5' untranslated region (5'UTR),
                      coding sequences (CDS, exon), intervening sequences
                      (intron), 3' untranslated region (3'UTR), and 3' clipped
                      region (3'clip)

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                                    
                      /evidence=&lt;evidence_value&gt;
                      /function="text"
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /standard_name="text"
                      /usedin=accnum:feature_label

<hr>
<a name="primer_bind">
Feature Key           primer_bind</a>


Definition            Non-covalent primer binding site for initiation of
                      replication, transcription, or reverse transcription.
                      Includes site(s) for synthetic e.g., PCR primer elements

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /standard_name="text"
                      /PCR_conditions="text"
                      /usedin=accnum:feature_label

Comment               This key is used to annotate the site on a given
                      sequence to which a primer molecule binds--it is not
                      intended to represent the sequence of the primer
                      molecule itself.
                      PCR components and reaction times may be stored under the
                      "/PCR_conditions" qualifier. 
                      Since PCR reactions most often involve pairs of primers,
                      a single primer_bind key may use the order() operator
                      with two locations, or a pair of primer_bind keys may be
                      used.

<hr>
<a name="promotor">
Feature Key           promoter</a>


Definition            region on a DNA molecule involved in RNA polymerase
                      binding to initiate transcription

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /function="text"
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /phenotype="text"
                      /pseudo
                      /standard_name="text"
                      /usedin=accnum:feature_label

Molecule scope        DNA

<hr>
<a name="protein_bind">
Feature Key           protein_bind</a>


Definition            non-covalent protein binding site on nucleic acid

Mandatory qualifiers  /bound_moiety="text"

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /function="text"
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /partial
                      /note="text"
                      /standard_name="text"
                      /usedin=accnum:feature_label

Comment               note that RBS is used for ribosome binding sites

<hr>
<a name="rbs">
Feature Key           RBS</a>


Definition            ribosome binding site


Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt; 
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /standard_name="text"
                      /usedin=accnum:feature_label

References            [1] Shine, J. and Dalgarno, L.  Proc Natl  Acad Sci USA
                          71, 1342-1346 (1974)
                      [2] Gold, L. et al.  Ann Rev Microb 35, 365-403 (1981)

Comment               in prokaryotes, known as the Shine-Dalgarno sequence: is
                      located 5 to 9 bases upstream of the initiation codon;
                      consensus GGAGGT [1,2]
                      /pseudo qualifier deleted v1.04

<hr>
<a name="repeat_region">
Feature Key           repeat_region</a>


Definition            region of genome containing repeating units


Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /function="text"
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /rpt_family="text"
                      /rpt_type=&lt;repeat_type&gt;
                      /rpt_unit=&lt;feature_label&gt; or &lt;base_range&gt;
                      /standard_name="text"
                      /usedin=accnum:feature_label

<hr>
<a name="repeat_unit">
Feature Key           repeat_unit</a>


Definition            single repeat element

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /function="text"
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /rpt_family="text"
                      /rpt_type=<repeat_type>
                      /rpt_unit=&lt;feature_label&gt; or &lt;base_range&gt;
                      /usedin=accnum:feature_label

Comment               preferred usage is to annotate the /rpt_family and
                      rpt_type qualifiers on the repeat_region, not on the
                      repeat_unit(s)

<hr>
<a name="rep_origin">
Feature Key           rep_origin</a>


Definition            origin of replication; starting site for duplication of
                      nucleic acid to give two identical copies 

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /direction=value
                      /evidence=&lt;evidence_value&gt;
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /standard_name="text"
                      /usedin=accnum:feature_label

Comment               /direction has valid values: RIGHT, LEFT, or BOTH

<hr>
<a name="rrna">
Feature Key           rRNA</a>


Definition            mature ribosomal RNA ; the RNA component of the
                      ribonucleoprotein particle (ribosome) which assembles
                      amino acids into proteins

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /function="text"
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /product="text"
                      /pseudo
                      /standard_name="text"
                      /usedin=accnum:feature_label

Comment               rRNA sizes should be annotated with the /product
                      qualifier.  It may be desirable to define a restricted
                      vocabulary for ribosomal RNA size classes.

<hr>
<a name="s_region">
Feature Key           S_region

 
Definition            Switch region of immunoglobulin heavy chains.  
                      Involved in the rearrangement of heavy chain DNA leading 
                      to the expression of a different immunoglobulin class 
                      from the same B-cell.

Optional qualifiers   /citation=[number]
	              /db_xref="<database>:<identifier>"
	              /evidence=<evidence_value>
	              /gene="text"
	              /label=feature_label
	              /map="text"
	              /note="text"
 	              /partial
                      /product="text"
                      /pseudo
                      /standard_name="text"
                      /usedin=accnum:feature_label

Parent Key            misc_signal

Organism scope        Eukaryotes

<hr>
<a name="satellite">
Feature Key           satellite</a>


Definition            many tandem repeats (identical or related) of a short
                      basic repeating unit;  many have a base composition or
                      other property different from the genome average  that
                      allows them to be separated from the bulk (main band)
                      genomic DNA

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /rpt_type=&lt;repeat_type&gt;
                      /rpt_family="text"
                      /rpt_unit=&lt;feature_label&gt; or &lt;base_range&gt;
                      /standard_name="text"
                      /usedin=accnum:feature_label

Molecule scope        DNA

Comment               use the satellite key to identify the entire region of
                      satellite sequence within an entry;  use repeat_unit to
                      identify individual repeated units (one is generally
                      sufficient) of the satellite;  

<hr>
<a name="scrna">
Feature Key           scRNA


Definition            small cytoplasmic RNA; any one of several small
                      cytoplasmic RNA molecules present in the cytoplasm and
                      (sometimes) nucleus of a eukaryote 

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /function="text"
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /product="text"
                      /pseudo
                      /standard_name="text"
                      /usedin=accnum:feature_label

<hr>
<a name="sig_peptide">
Feature Key           sig_peptide</a>


Definition            signal peptide coding sequence; coding sequence for an
                      N-terminal domain of a secreted protein; this domain is
                      involved in attaching nascent polypeptide to the
                      membrane; leader sequence

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /function="text"
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /product="text"
                      /pseudo
                      /standard_name="text"
                      /usedin=accnum:feature_label



<hr>
<a name="snrna">
Feature Key           snRNA</a>


Definition            small nuclear RNA; any one of many small RNA species
                      confined to the nucleus; several of the snRNAs are
                      involved in splicing or other RNA processing reactions

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /function="text"
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /product="text"
                      /pseudo
                      /standard_name="text"
                      /usedin=accnum:feature_label

<hr>
<a name="source">
Feature Key           source</a>


Definition            identifies the biological source of the specified span
                      of the sequence. This key is mandatory. Every entry will
                      have, as a minimum, a single source key spanning the
                      entire sequence. More than one source key per sequence
                      is permissable

Mandatory qualifiers  /organism="text"

Optional qualifiers   /cell_line="text"
                      /cell_type="text"
                      /chloroplast
                      /chromoplast
                      /chromosome="text"
                      /citation=[number]
                      /clone="text"
                      /clone_lib="text"
                      /cultivar="text"
                      /cyanelle="text"
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /dev_stage="text"
                      /focus="text"
                      /frequency="text"
                      /germline
                      /haplotype="text"
                      /lab_host="text"
                      /insertion_seq="text"
                      /isolate="text"
                      /kinetoplast
                      /label=feature_label
                      /macronuclear
                      /map="text"
                      /mitochondrion
                      /note="text"
                      /plasmid="text"
                      /pop_variant="text"
                      /proviral
                      /rearranged
                      /sequenced_mol="text"
                      /serotype="text"
                      /sex="text"
                      /specimen_voucher="text"
                      /specific_host="text"
                      /strain="text"
                      /sub_clone="text"
                      /sub_species="text"
                      /sub_strain="text"
                      /tissue_lib="text"
                      /tissue_type="text"
                      /transposon="text"
                      /usedin=accnum:feature_label
                      /variety="text"
		      /virion
		      
Molecule scope        any

Comment               multiple qualifiers (e.g., /clone=) may occur

<hr>
<a name="stem_loop">
Feature Key           stem_loop</a>


Definition            hairpin; a double-helical region formed by base-pairing
                      between adjacent (inverted) complementary sequences in a
                      single strand of RNA or DNA 

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /function="text"
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /standard_name="text"
                      /usedin=accnum:feature_label

<hr>
<a name="sts">
Feature Key           STS</a>


Definition            Sequence Tagged Site. Short, single-copy DNA sequence
                      that characterizes a mapping landmark on the genome and
                      can be detected by PCR. A region of the genome can be
                      mapped by determining the order of a series of STSs.

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /standard_name="text"
                      /gene="text"
                      /label=feature_label	
                      /usedin=accnum:feature_label
                      /note="text"
                      /partial
                      /map="text"

Molecule scope        DNA

Parent key            misc_binding

Comment               STS location to include primer(s) in primer_bind key or
                      primers.

<hr>
<a name="tata_signal">
Feature Key           TATA_signal</a>


Definition            TATA box; Goldberg-Hogness box; a conserved AT-rich
                      septamer found about 25 bp before the start point of
                      each eukaryotic RNA polymerase II  transcript  unit which
                      may be involved in positioning the enzyme  for correct 
                      initiation; consensus=TATA(A or T)A(A or T) [1,2]

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /usedin=accnum:feature_label

Organism scope        Eukaryotes and eukaryotic viruses

Molecule scope        DNA

References            [1] Efstratiadis, A.  et al.  Cell 21, 653-668 (1980)
                      [2] Corden, J., et al.  "Promoter sequences of
                          eukaryotic protein-encoding genes"  Science 209,
                          1406-1414 (1980)

<hr>
<a name="terminator">
Feature Key           terminator</a>


Definition            sequence of DNA located either at the end of the
                      transcript  that causes RNA polymerase to terminate 
		      transcription

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /standard_name="text"
                      /usedin=accnum:feature_label

Molecule scope        DNA

<hr>
<a name="transit_peptide">
Feature Key           transit_peptide</a>


Definition            transit peptide coding sequence; coding sequence for an
                      N-terminal domain of a nuclear-encoded organellar
                      protein; this domain is involved in post-translational
                      import of the protein into the organelle

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /function="text"
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /product="text"
		      /pseudo
                      /standard_name="text"
                      /usedin=accnum:feature_label


<hr>
<a name="trna">
Feature Key           tRNA</a>


Definition            mature transfer RNA, a small RNA molecule (75-85 bases
                      long) that mediates the translation of a nucleic acid
                      sequence into an amino acid sequence

Optional qualifiers   /anticodon=(pos:&lt;base_range&gt;,aa:&lt;amino_acid&gt;)
                      /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /function="text"
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /product="text"
                      /pseudo
                      /standard_name="text"
                      /usedin=accnum:feature_label

<hr>
<a name="unsure">
Feature Key           unsure</a>


Definition            author is unsure of exact sequence in this region

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /gene="text"
                      /usedin=accnum:feature_label
                      /label=feature_label
                      /map="text"
                      /note="text"

Comment               format of location is: 		
		      seq_location		/replace="change_loc"  
		      where change_loc is the sequence difference.  
                      Alternatively any valid location is acceptable and legal.


<hr>
<a name="v_region">
Feature Key           V_region</a>
 

Definition            Variable region of immunoglobulin light and heavy
                      chains, and T-cell receptor alpha, beta, and gamma
                      chains.  Codes for the variable amino terminal portion.
                      Can be made up from V_segments, D_segments, N_regions,
                      and J_segments.

Optional qualifiers   /citation=[number]
	              /db_xref="<database>:<identifier>"
	              /evidence=<evidence_value>
	              /gene="text"
	              /label=feature_label
	              /map="text"
	              /note="text"
 	              /partial
                      /product="text"
                      /pseudo
                      /standard_name="text"
                      /usedin=accnum:feature_label

Parent Key            CDS

Organism scope        Eukaryotes

<hr>
<a name="v_segment"> 

Feature Key           V_segment</a>


Definition            Variable segment of immunoglobulin light and heavy
                      chains, and T-cell receptor alpha, beta, and gamma
                      chains.  Codes for most of the variable region (V_region)
                      and the last few amino acids of the leader peptide

Optional qualifiers   /citation=[number]
	              /db_xref="<database>:<identifier>"
	              /evidence=<evidence_value>
	              /gene="text"
	              /label=feature_label
	              /map="text"
	              /note="text"
 	              /partial
                      /product="text"
                      /pseudo
                      /standard_name="text"
                      /usedin=accnum:feature_label

Parent Key            CDS

Organism scope        Eukaryotes

<hr>
<a name="variation">
Feature Key           variation</a>

Definition            a related strain contains stable mutations from the same
                      gene (e.g., RFLPs, polymorphisms, etc.) which differ
                      from the presented sequence at this location (and
                      possibly others).

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /frequency="text"
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /phenotype="text"
                      /product="text"
                      /standard_name="text"
                      /usedin=accnum:feature_label

Comment               format of location is: 		
		      seq_location		/replace="variation"  
		      where seq_location points to a region of the  presented sequence and 
		      variation is a location descriptor (typically a literal sequence) with 
		      which the sequence at seq_location is replaced in the variant.
                      Alternatively any valid location is acceptable and legal.

<hr>
<a name="3clip">
Feature Key           3'clip</a>


Definition            3'-most region of a precursor transcript that is clipped
                      off during processing

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /function="text"
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /standard_name="text"
                      /usedin=accnum:feature_label

<hr>
<a name="3utr">
Feature Key           3'UTR</a>


Definition            region at the 3' end of a mature transcript (following the stop codon) 
		      that is not translated into a protein.

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /function="text"
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /standard_name="text"
                      /usedin=accnum:feature_label

<hr>
<a name="5clip">
Feature Key           5'clip</a>


Definition            5'-most region of a precursor transcript that is clipped
                      off during processing

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /function="text"
                      /gene="text"
                      /evidence=&lt;evidence_value&gt;
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /standard_name="text"
                      /usedin=accnum:feature_label

<hr>
<a name="5utr">
Feature Key           5'UTR</a>


Definition            region at the 5' end of a mature transcript (preceding the initiation 
		      codon) that is not translated into a protein.

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /function="text"
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /standard_name="text"
                      /usedin=accnum:feature_label

<hr>
<a name="10_signal">
Feature Key           -10_signal</a>


Definition            Pribnow box; a conserved region about 10 bp upstream of
                      the start point of bacterial transcription units which
                      may be involved in  binding RNA polymerase;
                      consensus=TAtAaT [1,2,3,4]

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /partial
                      /note="text"
                      /standard_name="text"
                      /usedin=accnum:feature_label

Organism scope        prokaryotes

Molecule scope        DNA

References            [1] Schaller, H., Gray, C., and Hermann, K.  Proc Natl
                          Acad Sci USA 72, 737-741 (1974)
                      [2] Pribnow, D.  Proc Natl Acad Sci USA 72, 784-788 (1974)
                      [3] Hawley, D.K. and McClure, W.R.  "Compilation and
                          analysis of Escherichia coli promoter DNA sequences" 
                          Nucl Acid Res 11, 2237-2255 (1983)
                      [4] Rosenberg, M. and Court, D.  "Regulatory sequences
                          involved in the promotion and termination of RNA
                          transcription"  Ann Rev Genet 13, 319-353 (1979)

<hr>
<a name="35_signal">
Feature Key           -35_signal</a>


Definition            a conserved hexamer about 35 bp upstream of the start
                      point of bacterial transcription units; consensus=TTGACa
                      [  ] or TGTTGACA [  ]

Optional qualifiers   /citation=[number]
                      /db_xref="&lt;database&gt;:&lt;identifier&gt"                      
                      /evidence=&lt;evidence_value&gt;
                      /gene="text"
                      /label=feature_label
                      /map="text"
                      /note="text"
                      /partial
                      /standard_name="text"
                      /usedin=accnum:feature_label

Organism scope        prokaryotes

Molecule scope        DNA

References            [1] Takanami, M., et al.  Nature 260, 297-302 (1976)
                      [2] Moran, C.P., Jr., et al.  Molec Gen Genet 186,
                          339-346 (1982)
                      [3] Maniatis, T., et al.  Cell 5, 109-113 (1975)

Comment               note: original signal sequence proposed was TGTTG [4]
</pre>
