%
% Philosophy_gff2ps.tex
%
% Main points on ideas behind ``gff2ps''.
%
% $Id: Philosophy_gff2ps.tex,v 1.1 2002-12-24 13:07:37 jabril Exp $
%

The programming philosophy underlying \prog\ can be summarized onto these points:
%\begin{minipage}[b][16cm][c]{14cm}

\begin{figure}[!ht]
\begin{center}
	\input{Blocks}
	\caption[Plot distribution for elements found in GFF records]{Here, you can observe in upper Block how strands are distributed on block area, and also how sources have reverse order in reverse strand. You can disable visualization for any strand and place one or more blocks per page. Imagine that lower Block has the same distribution as upper one, but the figure is only showing one zoomed strand. That strand also focuses on one source track, what is represented on it, and how we represent grouped and un-grouped elements generated from GFF-features. There is also shown overlapped elements and groups, each of them can be treated in different ways by \prog.}
	\label{plothierarchy} 
\end{center}
\end{figure}

%
%\hspace{-2.5cm}
\begin{figure}[!ht]
\begin{center}
\fbox{
\begin{minipage}[!ht][][c]{\linewidth}\scriptsize\ttfamily
\begin{center}
\begin{tabular}{ccc}
\textbf{\normalsize GFF Pattern}                  &               & \textbf{\normalsize Group Examples} \\ \hline\hline
\\
\symbol{91}1--8\symbol{93} .$\ast$  \symbol{91}\ldots\symbol{93}   & $\Rightarrow$ & $\left\{\begin{array}{c}\mbox{123}\\\mbox{DMSELE.1}\\\mbox{Clone\_33223}\end{array}\right.$ \\
\\
\symbol{91}1--8\symbol{93} ".$\ast$" \symbol{91}\ldots\symbol{93}  & $\Rightarrow$ & $\left\{\begin{array}{c}\mbox{"123"}\\\mbox{"DMSELE.1"}\\\mbox{"Clone\_33223"}\\\mbox{"Brain K+ Channel"}\end{array}\right.$ \\
\\
\symbol{91}1--8\symbol{93} \symbol{91}A-Za-z\symbol{93}\symbol{91}A-Za-z0-9\_\symbol{93}$\ast$ ".$\ast$" \symbol{91}\ldots\symbol{93} & $\Rightarrow$ & $\left\{\begin{array}{c}\mbox{target "123"}\\\mbox{SIMILARITY "DMSELE.1"}\\\mbox{label "Clone\_33223"}\\\mbox{Putative\_Protein "Brain K+ Channel"}\end{array}\right.$ \\
\end{tabular}
\end{center}
\end{minipage}}
\noindent
\caption[Group formats read by \prog.]{Group formats read by \prog. `\texttt{\symbol{91}1--8\symbol{93}}' represents the first eight gff-fields of each GFF-record. `\symbol{91}\ldots\symbol{93}' corresponds to extra fields that are not used by our program.}
\label{valid_groups}
\end{center}
\end{figure}
%

\begin{itemize}
%\setlength{\parsep}{0ex plus0ex}
%\setlength{\itemsep}{0ex plus0ex}
%\setlength{\topsep}{0ex plus0ex}
%\setlength{\partopsep}{0ex plus0ex}
\item[$\bullet$] We want to generate comprehensive plots of all `GFF-able' features in order to compare genomic sequences from different sources. Although developed initially to display features annotated from different sources on a single sequence, it can also be used for displaying annotations from one (or more) sources on a number of sequences. This can be useful, for instance to compare the genomic structure of different sequences.
\item[$\bullet$] \prog\ can parse Version 1 and Version 2 GFF-formatted records, records that are not compliant with GFF-format are discarded warning user afterwards. Field separator and group field were defined slightly different from one version to the other, here it is explained how \prog\ deals with those differences.
If records from input GFF-files contain tabulators as field separator program assumes that record is GFF Version 2 formatted, else if it finds blank-spaces as field separator then switches to Version 1.\\
Groups must be defined as a `tag-value' pair in GFF Version 2, where `tag' must be a standard identifier (\symbol{91}A-Za-z\symbol{93}\symbol{91}A-Za-z0-9\_\symbol{93}$\ast$) and `value' must be a free-text string enclosed between double-quotes (`\,\texttt{"}.$\ast$\texttt{"}\,'), for example `\texttt{target "HS new gene"}' fits that group format. Other group formats force program to switch to GFF Version 1: if there are more than nine fields it tries to find the `tag-value' pattern, else assumes that the ninth field is the group name as a quoted or not free-text string. See figure~\ref{valid_groups} for group formats than can be processed by \prog.\\
%\item[$\bullet$] \prog\ Works with GFF Version 2 formatted lines, that implies that records are checked in a slightly different way to Version 1, mainly due to field separator constraint and group field definition. Whereas the group can be used to set groups on the plot, we recommend to define it as a label not as a fixed tag (we are working for `target' and `transcript' following labels will be also used for grouping the gff elements), see figure~\ref{GFFgroups} on section~\ref{sec:philoGFF} for some grouping examples. Records that are not compliant with GFF-format are discarded, warning user for that and not used by \prog.
\item[$\bullet$] The program must be easy to use, all its parameters are set by default inside the program. But must be easy to change plot options which can be modified by a default custom file, also by a working custom file (smaller than default file and provided to introduce small changes for single plots), and some of them from command-line. The main goal is to define a system in which can be easily added new options or redefine old ones. Another issue is that configuration files are plain-ASCII text, so they can be edited with small and simple text editors. To know more about this issue you can read section~\ref{sec:VdefCF} and~\ref{sec:CustFiles}. The program can work in background or used in a \textsc{Unix} pipeline working as a filter (section~\ref{sec:unixCL}).
\item[$\bullet$] Source order from input gff-file is preserved when reading those files, it means that you can easily switch source order in your plot swapping the order of the input files. Sources are shown in plots giving a mirroring symmetry axes for strands forward and reverse, so forward sources are shown by its ordering from top to bottom and reverse sources are shown counter-wise, while records without a defined strand are placed in the plot area between the two strands areas.
\item[$\bullet$] User-defined custom files can handle regular expressions, allowing us to define any attribute variable for multiple similar features ---in GFF-elements, group or source blocks--- in one line, to know about this feature you can look at section~\ref{sec:UsingREGEXP}.
\item[$\bullet$] Some of the previous items leads to hierarchical plots, in which Pages are the highest element, and Blocks are defined within them ---this feature allows you to get multiple vertical and/or horizontal pages, also pages with multiple blocks---. Inside Blocks may appear any Strand ---forward, reverse, or no-frame (defined in GFF-records as `.')---. Each Strand presents each Source as one plot line or more ---if you want to display overlapping groups in the same line and/or in different lines or if you prefer to split data-sets between grouped and un-grouped features---, meanwhile Groups belonging to them define features for displaying sets of GFF-features, which are the basic plot elements (schemed in figure~\ref{plothierarchy}). Page number and Blocks per page are set as a Layout variables (see section~\ref{sec:layoutft}), whereas the rest of other elements are defined in specific fields from GFF-files records: Sources are named at second field, GFF-features came from third field, Strands from field seven, and Groups from ninth (see section~\ref{sec:GFFdefs} and tables~\ref{GFFrecords} and \ref{GFFgroups}). Start, End, Score and Frame are defined into GFF-features as plot attributes. 
\item[$\bullet$] You can switch on/off visualization of overlapping groups in several lines for same source track ---option by default---, or you can fit them into a single track. \prog\ minimizes the number of lines needed for that when displaying overlapping groups in different lines. Individual non-grouped GFF-elements are treated as a one element group, which allows you to display also individual elements without overlapping. If you choose to print in a single line, you can also define layers for each set of overlapping GFF-features, maybe `exons' at top and `cds' at bottom, enhancing viewing for any of the elements. See sections~\ref{sec:gffelemft} and~\ref{sec:groupft}.
\item[$\bullet$] You must remember that \prog\ converts all upper-case characters for \textbf{features} to lower-case. In order to prevent that one user had defined `Exon', another `exon' and other `EXON', our program convert them to `exon'.
\item[$\bullet$] Scores control feature width, but in order to prevent problems when working with data-sets provided by different programs they are re-scaled for each source, using maximum and minimum values as a score range within all scores are re-calculated. Default score is set to maximum value for each source (when parsing gff-files and a record contains a `.' in Score field). Further information can be found in section~\ref{sec:sourceft}.
\item[$\bullet$] We have defined three block areas where are displayed Strands, you can switch on/off any of them and visualize one up to three strands. Forward strand (`+') is always shown at upper area, while reverse strand is always shown at lower area. When strand is not defined (`.') elements are placed in the central area, between forward and reverse areas. Source order is preserved from input files, in forward strand sources follows that ordering up to down, also in the no-strand area, but in reverse strand are shown down to up, so you have a horizontal symmetry axes between forward/no-strand and reverse strand.
\item[$\bullet$] Features for which frame is specified are plotted using a two color code schema. The upstream half of the graphical element representing the frame of feature and the downstream half the complement modulus three of its remainder. This is useful to check frame consistency between adjacent features (for instance, predicted exons). Two adjacent features are frame-compatible when the color of the downstream half of the upstream feature matches the color of the upstream half of the downstream feature. This two-color code schema, however, is only meaningful when the frame has been defined relative to the feature, and not relative to the sequence. We have defined four independent frames in \prog (`.', `0', `1', and `2'), that is used by a coloring procedure for visualizing frame and remainder within shapes. Colors for frames are also independently defined from feature color definition, and can be customized too. The complement modulus three for ``remainder'' is calculated by our program following this formula:\\[1ex]
\centerline{\label{fml:remainder}$\mbox{\textsl{``Remainder''}} = ( 3 - ( \mbox{\textsl{End}} - ( \mbox{\textsl{Start}} + \mbox{\textsl{Frame}} ) + 1 ) \bmod 3 ) \bmod 3$}
\item[$\bullet$] Score vectors are shown as score-dependent color gradients and are read from one line which contain such vectors in those fields beyond group field. We will implement in next version procedures for showing them as continuous or discrete functions, spikes, and so on... You must use the following format for the group and the following fields (next version will work also with multiple GFF score-feature single-records) :\label{vectordef}\\

\centerline{\small\symbol{91}Fields 1 to 8\symbol{93}\ \ Tag\ ``Value''\ \ \textbf{score};\ \ \textbf{Window} \textit{window};\ \ \textbf{Step} \textit{step};\ \ \textbf{Scores} \textit{score} \symbol{91}\ldots\symbol{93} \textit{score}}

\end{itemize}



